## Molecular epidemiology and genomic analysis of individual patient isolates and environmental sampling reveals transmission of bacterial plasmids harboring carbapenem resistant genes

[RefPaper1](https://sci-hub.se/https://doi.org/10.1093/jac/dkaa398)
### Introduction:
* Most important problem in healthcare setting
* Which order of bacteria is causing it? What makes it more deadly and powerful to attack the humans?
* CPE - mostly which organisms cause the frequent attacks?
* What causes the Carbapenemase to move? Transmissible mobile genetic elements
* Which sepcies in this study we are concentrating about?
* Are there any prevalent ST clone in our dataset?
* Treatment of antibiotics
* Acquisition of resistance genes through mobile genetic elements
    - specific genomic islands 
    - transposons
    - plasmids
    - integrative and conjugative elements
    - integrons
* Severe infection due to virulence factors
* Toxins of the type III secretion system to circumvent the host immune system and establish infection
* Of the four exotoxins (ExoS, ExoT, ExoU, ExoY), ExoU, a potent phospholipase that disrupts the plasma membrane and leads to rapid cell death, is the most virulent 

### Methods

### Results

* **Bacterial isolates**
  - From what time to what time these isolates have been collected?
  - Out of the collected, how many have carbapenemases?
  - Where are isolates collected from patients? Which body sites? If environmental, which surrounding sites? Give the count of each site.
  - Out of this, what bacterial species was found and the count of each species.
  - Is there any trend of increasing in the number of isolates by species by year?

  
* **Location of CP genes**
  - Is it chromsomosomal or plasmids?
  - If plasmids what is the size of the plasmids

* **WGS analysis**
   - Genome Characteristics: The average genome size of the de novo assembled genomes for each species - Information about all genome characteristics is summarized in Table. 
   - Resistome analysis - what genes are conferring resistance to aminoglycosides (aacA4 and aadA1), macrolides (mphE and msrE), chloramphenicol (catA1 and catB2), sulphonamides (sul1) and trimethoprim/sulfamethoxazole (dfrB1) in each of the species isolates. In OXA-48-CP isolates, only one gene involved in aminoglycoside antimicrobial resistance was found (aacA4) (Figure S3). Show the facet grid or PCA plot for the resistome
   - Type of plasmid reconstructions showed in each of the 4 species - what is the size of the plasmid with CP gene ? Is it available in NCBI? If not, what is the coverage and identity? Is this the same plasmid in the other species as well carrying the CP gene? What are other plasmids in other species? and their details?
   - Phylogenetic analysis of all the genomes for each species 
   - Typing, plasmid, and genomic transmission analysis - What does MLST analysis reveal? How many groups of STs identified for each species? 
   - Plasmid replicon types were detected, and associations with carbapenemase genes were assessed - Among OXA-48-producing isolates, Inc-- replicon plasmids were common in Serratia marcescens (85.7%) and K. pneumoniae (61.1%), IncL/M plasmids were more common among E. coli and Enterobacter spp. (66.7% and 50.0%, respectively), and IncHI2 plasmids were most common among E. cloacae complex (61.5%). All three types have previously been described to carry the IMP-4 gene in Australia (9, 34). KPC-2-producing K. pneumoniae commonly carried IncFIB (94.6%), ColRNAI (90.2%), and IncX3 (85.7%). In contrast, NDM and OXA isolates carried a wide range of plasmid types. Plasmid replicon sequences were only rarely detected on the same contig as a carbapenemase gene. Fifteen isolates had no plasmid replicon sequences detected at these identity/coverage cutoffs, including all IMI- and SME-producing isolates (chromosomal carbapenemases).
   - Genomic transmission analyses were performed on 33 subgroups of the same species, carbapenemase gene, and ST (where available), including 131 isolates from 119 patients; subgroups included 2 to 14 isolates (median, 3 isolates) (Table S6). Pairwise SNP distances were plotted for each species
   - Plasmid replicon typing results were consistent with transmission analysis results in epidemiologically confirmed clusters, in that all isolates within a cluster carried the same plasmid replicon types, with the exception of the mixed NDM-5/NDM-5 plus OXA-232 cluster (K. pneumoniae ST16), where only the isolates with OXA-232 had the ColKP3-type plasmid detected, suggesting carriage of the OXA-232 gene on this plasmid.

### Discussion
   - Resistome analysis




* **ST308 population structure by phylogenetic network analysis:** 
    - How many clades?
    - largest clade? 
    - Are the isolates between the country/regions/hospital scattered in all clades?
    - Are there any clades specific to a country? 
    - Do the clades have specific geographical origin? 
    - Any clades with all the country/regions/hospital?
    - 
    
* **Spatiotemporal origin of the ST308 clone**
    - What is tMRCA of the isolates date to? What is the HPD interval from the date of isolation of the latest isolate in year 2020?
    - When was the oldest isolate isolated and from where country/hospital/continent?
    - When does clade show that the isolate was emerged in the country? Is the split in the year 2005 confirmed with Bayesian approach with a posterior probability score of 1.0. This suggests that in Singapore the spread of ST308 is comprised of atleast two independent clades.
    - Are the SNPs present in the core genome? How much percentage of the SNPs are present in resistant genes and how many in virulent genes?
    - When is the emergence of year of different groups and clades?
    - tMRCA tree figure 

* **High diversity of foreign antibiotic resistant determinants among the ST308 isolates**
    - Resistome Figure, a table in supplementary
    - Prevalence of antibiotics figure
    - How much percentage of isolates possess how many classes of ARG?
    - How much percentage of isolates possess how many classes of Beta-lactamases?    
    - How much percentage of isolates possess how many classes of Carbapenemases? What are they? Any other types of carbapenemases?
    - Any specific clades with specific carbapenemases?
    - The number of acquired resistance genes was significantly lower or higher in the isolates collected between two clades (Inst A and Inst B)
   
 * **ST308-specific determinants**
    -  Does St308 rely on any specific determinants?
    -  Is the gene signature shared among the widespread clones?
    -  Any specific genes found only in the ST308 isolates but not nonST308 clones?

### Discussion

 * **ST308-specific determinants**
    -  Does ST308 rely on any specific determinants? How about the genes DprA, RecQ, RecA, - needs more study 

 * **Not much diversity of resistance genes in ST308 isolates**
    - Spread of both of these lineages are more or less heavily dependent on these resistant genes

 * **Role of Fluoroquinolones**
    - Overall, although other features of ST235 may have contributed to its spread, fluoroquinolone use could have highly favoured the spread of exoU strains (including ST308) in which the fitness burden of the resistance to these major antibiotics is lower.
 
 
 
 
 ## Additional backup
 
 Also can write up if the samples are specific to any ST types
 
#### ST clone

* ST235 one of the high risk clone and poorly understood

#### History of the ST308 clone

* To better understand the history of the P. aeruginosa STXXX lineage as a high-risk international clone, we carried out a Bayesian phylogenetic reconstruction using XXX P. aeruginosa STXXX isolates collected from five continents/countries/hospitals over a XXX-year period.

#### Genomic study of the ST308 clone

* Genome comparison of these isolates identified antibiotic-resistance determinants, virulence genes and ST235-specific genes that may have contributed to the success of this clone.
