### Introduction:
* Most prevalent ST clone
* Treatment of antibiotics
* Acquisition of resistance genes through mobile genetic elements
    - specific genomic islands 
    - transposons
    - plasmids
    - integrative and conjugative elements
    - integrons
* Severe infection due to virulence factors
* Toxins of the type III secretion system to circumvent the host immune system and establish infection
* Of the four exotoxins (ExoS, ExoT, ExoU, ExoY), ExoU, a potent phospholipase that disrupts the plasma membrane and leads to rapid cell death, is the most virulent 

#### ST clone

* ST235 one of the high risk clone and poorly understood

#### History of the ST308 clone

* To better understand the history of the P. aeruginosa STXXX lineage as a high-risk international clone, we carried out a Bayesian phylogenetic reconstruction using XXX P. aeruginosa STXXX isolates collected from five continents/countries/hospitals over a XXX-year period.

#### Genomic study of the ST308 clone

* Genome comparison of these isolates identified antibiotic-resistance determinants, virulence genes and ST235-specific genes that may have contributed to the success of this clone.

### Methods

### Results

* **ST308 population structure by phylogenetic network analysis:** 
    - How many clades?
    - largest clade? 
    - Are the isolates between the country/regions/hospital scattered in all clades?
    - Are there any clades specific to a country? 
    - Do the clades have specific geographical origin? 
    - Any clades with all the country/regions/hospital?
    - 
    
* **Spatiotemporal origin of the ST308 clone**
    - What is tMRCA of the isolates date to? What is the HPD interval from the date of isolation of the latest isolate in year 2020?
    - When was the oldest isolate isolated and from where country/hospital/continent?
    - When does clade show that the isolate was emerged in the country? Is the split in the year 2005 confirmed with Bayesian approach with a posterior probability score of 1.0. This suggests that in Singapore the spread of ST308 is comprised of atleast two independent clades.
    - Are the SNPs present in the core genome? How much percentage of the SNPs are present in resistant genes and how many in virulent genes?
    - When is the emergence of year of different groups and clades?
    - tMRCA tree figure 

* **High diversity of foreign antibiotic resistant determinants among the ST308 isolates**
    - Resistome Figure, a table in supplementary
    - Prevalence of antibiotics figure
    - How much percentage of isolates possess how many classes of ARG?
    - How much percentage of isolates possess how many classes of Beta-lactamases?    
    - How much percentage of isolates possess how many classes of Carbapenemases? What are they? Any other types of carbapenemases?
    - Any specific clades with specific carbapenemases?
    - The number of acquired resistance genes was significantly lower or higher in the isolates collected between two clades (Inst A and Inst B)
   
 * **ST308-specific determinants**
    -  Does St308 rely on any specific determinants?
    -  Is the gene signature shared among the widespread clones?
    -  Any specific genes found only in the ST308 isolates but not nonST308 clones?

### Discussion

 * **ST308-specific determinants**
    -  Does ST308 rely on any specific determinants? How about the genes DprA, RecQ, RecA, - needs more study 

 * **Not much diversity of resistance genes in ST308 isolates**
    - Spread of both of these lineages are more or less heavily dependent on these resistant genes

 * **Role of Fluoroquinolones**
    - Overall, although other features of ST235 may have contributed to its spread, fluoroquinolone use could have highly favoured the spread of exoU strains (including ST308) in which the fitness burden of the resistance to these major antibiotics is lower.
 
 
