---
title: "5 year CPE Tranmission - Gene Intersection across species by chromosomes and plasmids"
author: "Prakki Sai Rama Sridatta"
date: "`r format(Sys.Date())`"
output:
  html_document:
    toc: TRUE
    toc_float: TRUE
    code_folding: hide
    keep_md: true
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
#knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning = FALSE, message = FALSE) 
```

#### Ran the below commands in Unix shell

```
cd /data02/Analysis/Projects/2_CPE_Transmission/VennDiagram_of_genes_between_species/CPE_Transmission_Data_Analysis


#---------- Identity plasmids and chromosomes using PlasClass

time for j in $(ls *.fasta); do echo $j; time python3.7 ~/sw/PlasClass/classify_fasta.py -f $j -o "$j".plasclass.probs.out -p 60;  done
real	464m57.242s
user	592m31.525s
sys	133m29.767s

mkdir ../1_plasClass_Probs
mv *.plasclass.probs.out ../1_plasClass_Probs
$ awk '{print FILENAME "\t" $0}' *.out >CPE_Trans_PlasClass_Predictions_comb.tab



#---------- Identify Transposons using Bacant (Integron module is not working properly and Resistance genes are used from AMRfinderplus)
time for d in $(ls *.fasta);  do  
echo "$d";
bacant -n "$d" -o "$d"_bacant_out &
done

mkdir ../2_bacant_annot
mv *_bacant_out ../2_bacant_annot

awk '{$13=$14=""; print FILENAME "\t" $1 "\t" $2}' *_bacant_out/transposon.filter.tsv | fgrep -v 'QUERY' | sed 's/_bacant_out\/transposon.filter.tsv//g' | sed -e 's/_circ[0-9]*/_circ/g' -e 's/\(depth_[0-9]*.[0-9]*x\)[0-9]*/\1/g' | awk '{print $1 "\t" $2 "\t" $1"#"$2 "\t" $3}' >CPE_Trans_BacAnt_Transposons.tab 

awk '{print FILENAME "\t" $0}' *_bacant_out/integron.filter.tsv | fgrep -v 'QUERY' | sed 's/_bacant_out\/integron.filter.tsv//g' | sed 's/[0-9]*\~integron_..//g' | awk '{print $1 "\t" $2 "\t" $1"#"$2 "\t" $3}' >CPE_Trans_BacAnt_Integrons.tab

#---------- Identify IS using Prokka (NOTE: I used an updated ISdb (default ISdb for prokka) located in my desktop)
time awk '{print FILENAME "\t" $0}' *_prokka_out/*.gff | grep 'ISfinder'  | sed 's/ /_/g' | awk '{print $1 "\t" $2 "\t" $NF}' | sed -e 's/ID=.*ISfinder://g' -e 's/;locus_tag.*//g' -e 's/html.2020\/\///g' | sed 's/_prokka_out.*gff/_assembly_renamed.fasta/g' | awk '{print $1 "\t" $2 "\t" $1"#"$2 "\t" $3}' >../../VennDiagram_of_genes_between_species/CPE_Transmission_Data_Analysis/6_Prokka_InsertionSequences/CPE_Trans_Prokka_IS.tab 



#---------- Identity AMR genes using NCBI AMRfinderplus

# NOTE: Ran this earlier, so just copying the results

mkdir ../3_amrfinderplus_results 

#---------- Identity Virulence factor genes using Abricate

for d in $(ls *.fasta); do echo "abricate -db vfdb $d >"$d".vfdb.tab"; done >abricate_cmds.sh
time parallel --jobs 62 < abricate_cmds.sh &

mkdir ../4_abricate_vf_results 
mv *.vfdb.tab ../4_abricate_vf_results
cd ../4_abricate_vf_results

cat *_assembly_renamed.fasta.vfdb.tab | fgrep -v '#' | awk '{print $1 "\t" $2 "\t" $3 "\t" $4 "\t" $5 "\t" $6}' >CPE_Trans_Abricate_VF_Genes.tab 


## Excluding possible contamination samples
-rwxrwxr-x 1 prakki prakki 9.8M Jul  6 14:17 batch12_27072019_ENT1410_assembly_renamed.fasta
-rwxrwxr-x 1 prakki prakki  11M Jul  6 14:17 batch2_21032019_ENT466_assembly_renamed.fasta
-rwxrwxr-x 1 prakki prakki  11M Jul  6 14:17 batch2_21032019_ENT497_assembly_renamed.fasta
-rwxrwxr-x 1 prakki prakki  11M Jul  6 14:18 batch2_21032019_ENT378_assembly_renamed.fasta
-rwxrwxr-x 1 prakki prakki  11M Jul  6 14:18 rbatch2_07082020_ENT792_assembly_renamed.fasta

# These assemblies have abnormal assembly sizes. So excluding them from analysis

## -- Reunning MLST 

for d in $(ls *.fasta); do echo "mlst $d >"$d".mlst.tab"; done >mlst_cmds.sh
time parallel --jobs 62 < mlst_cmds.sh &

```

#### Ran the below commands in R



```{r geneSharing, fig.width=18, fig.height=10, echo=TRUE,message=FALSE}
library(dplyr)
library(UpSetR)
library(tidyverse)
library(splitstackshape)
library(reshape2)
library(randomcoloR)
library(ComplexHeatmap)
library(data.table)
library(kableExtra)
library(DT)
library(data.table)
library(formattable)




setwd("/data02/Analysis/Projects/2_CPE_Transmission/VennDiagram_of_genes_between_species/CPE_Transmission_Data_Analysis")


hybridAssemb_1198list <- readLines("/data02/Analysis/Projects/2_CPE_Transmission/VennDiagram_of_genes_between_species/CPE_Transmission_Data_Analysis/CPE_Hyrbid_Assemblies_1198.list")
#head(hybridAssemb_1198list)

###@@@@@@@@@@@@@@ Dataframe1 - PlasClass Predictions

plasClass_Prob_df <- data.table::fread(file = "1_plasClass_probs/CPE_Trans_PlasClass_Predictions_comb.tab", sep = "\t", header = FALSE)
#head(plasClass_Prob_df)

colnames(plasClass_Prob_df) <- c("Fasta", "Contig", "Probability")

plasClass_Prob_df <- 
  plasClass_Prob_df %>%  
  mutate(Fasta = stringr::str_replace(Fasta, ".plasclass.probs.out", "")) %>% 
  mutate(Fasta_Contig=paste(Fasta, Contig, sep = '#')) %>% 
  mutate(Classification = if_else(Probability >= 0.6, "Plasmid", "Chromosome")) 

#head(plasClass_Prob_df)

###@@@@@@@@@@@@@@ Dataframe2 - MLST

mlst_df <- data.table::fread(file = "5_MLST/mlst_log2.19", sep = "\t", header = FALSE, fill = TRUE)
#head(mlst_df) # mlst_df now has 1289 assembly information
colnames(mlst_df) <- c("Fasta", "Species", "ST", "Gene1", "Gene2", "Gene3", "Gene4", "Gene5", "Gene6", "Gene7")
mlst_df <- mlst_df %>% filter(Fasta %in% hybridAssemb_1198list) # mlst_df now has only 1198 assembly information

Species_Counts <- as.data.frame(table(mlst_df$Species))
colnames(Species_Counts) <- c("Species","SampleCount")

Species_Counts %>% 
  kbl(caption = "Species and Sample Counts") %>%
  kable_classic(full_width = F, html_font = "Cambria")
#head(Species_Counts)

#----------------Step1: Combining MLST and plasClass dataframes

plasClass_Prob_df <- left_join(plasClass_Prob_df, mlst_df, by = c("Fasta")) %>% 
    select(Fasta,Contig,Fasta_Contig,Classification,Species, ST) %>% 
    mutate(Species_ContigClass=paste(Species,Classification, sep = '#'))

#head(plasClass_Prob_df)

###@@@@@@@@@@@@@@ Dataframe2 - Bacant Predictions Transposons

bacAnt_df <- data.table::fread(file = "2_bacant_annot/CPE_Trans_BacAnt_Transposons.tab", sep = "\t", header = TRUE)
#head(bacAnt_df)

colnames(bacAnt_df) <- c("Fasta", "Contig", "Fasta_Contig", "Transposon")

#----------------Step2: Combining plasClass dataframe with bacant dataframe for plotting Transposons

bacant_plasClass_df <- left_join(bacAnt_df, plasClass_Prob_df, by = c("Fasta_Contig")) %>% 
  select(Fasta.x,Contig.x,Fasta_Contig,Transposon,Classification,Species_ContigClass)

#head(bacant_plasClass_df)

bacant_plasClass_df <- bacant_plasClass_df %>% filter(Fasta.x %in% hybridAssemb_1198list) # bacant_plasClass_df now has only 1198 assembly information

# bacant_plasClass_df %>%
#   select(Transposon,Species_ContigClass) %>%
#   table() %>% # frequency of unique transposons
#   kbl(caption = "Transposons and their counts in Chromosomes and Plasmids across Species") %>%
#   kable_classic(full_width = F, html_font = "Cambria")

customGreen0 = "#DeF7E9"
customRed = "#ff7f7f"

bacant_plasClass_df_total_col = bacant_plasClass_df %>%
  select(Fasta.x,Transposon,Species_ContigClass) %>%
  unique() %>%
  select(Species_ContigClass) %>% 
  unique() %>% 
  drop_na(Species_ContigClass) %>% count() %>% c()
```

## Transposons {.tabset}

### Table1

#### Percentages of Samples carrying transposons in Chromsomes and Plasmids across Species

```{r geneSharing_transposons_samplepercentage, fig.width=18, fig.height=10, echo=TRUE,message=FALSE}
# percentages of transposons (https://clarewest.github.io/blog/post/making-tables-shiny/)
bacant_plasClass_df %>%
  select(Fasta.x,Transposon,Species_ContigClass) %>%
  unique() %>%
  select(Transposon,Species_ContigClass) %>%
  count(Transposon, Species_ContigClass) %>%
  rename(Freq = n) %>%
  drop_na(Species_ContigClass) %>%
  mutate(Species_ContigClass2=Species_ContigClass) %>%
  separate(Species_ContigClass2, sep = "#", into = c("Species", "ContigClass")) %>%
  left_join(Species_Counts, by = c("Species")) %>%
  mutate(SamplePrcnt_wGeneElement=Freq*100/SampleCount) %>%
  select(Transposon,Species_ContigClass,SamplePrcnt_wGeneElement) %>%
  data.table::dcast(Transposon~Species_ContigClass, value.var="SamplePrcnt_wGeneElement") %>%
  replace(is.na(.), 0) %>%
  mutate_if(is.numeric, round, digits=1) %>% 
  formattable(., align = c("l",rep("c", bacant_plasClass_df_total_col$n)), 
              list(
    `Transposon` = formatter("span", style = ~ style(color = "grey",font.weight = "bold")), 
    area(col = 1:bacant_plasClass_df_total_col$n+1) ~ color_tile(customGreen0, customRed))) %>% 
  as.datatable(escape = FALSE,
               options = list(scrollX = FALSE, dom = 'lfti',
                              lengthMenu = list(c(5, 10, 25, 100, -1), c('5','10','25', '100', 'All')),
                              pageLength = 5),
               rownames = FALSE)
  
#head(bacant_plasClass_df)
```

### Table2

#### Transposons and their unique counts in Chromsomes and Plasmids across Species

```{r geneSharing_transposons_count, fig.width=18, fig.height=10, echo=TRUE,message=FALSE}
bacant_plasClass_df %>% 
  select(Transposon,Species_ContigClass) %>% 
  unique() %>% 
  table() %>% # frequency of unique transposons
  as.data.frame.matrix() %>% 
  kbl(caption = "Transposons and their unique counts in Chromsomes and Plasmids across Species") %>% 
  kable_classic(full_width = F, html_font = "Cambria") 
```

### Transposons UpsetR 

#### UpsetR Plot Transposons

```{r geneSharing_transposons_UpsetR, fig.width=18, fig.height=10, echo=TRUE,message=FALSE}
#----------------Step3: UpsetR Plot Transposons
 
tn_upsetR_df <- bacant_plasClass_df %>% select(Transposon,Species_ContigClass)
#View(tn_upsetR_df)
tn_upsetR_lt <- split(tn_upsetR_df$Transposon, tn_upsetR_df$Species_ContigClass) 
tn_comb_mat = make_comb_mat(tn_upsetR_lt)

#tn_comb_mat
#str(tn_comb_mat)

#as.data.frame(tn_comb_mat)
# comb_mat
# UpSet(tn_comb_mat)
# UpSet(t(comb_mat))
# UpSet(comb_mat, top_annotation = upset_top_annotation(comb_mat, add_numbers = TRUE),
# right_annotation = upset_right_annotation(comb_mat, add_numbers = TRUE))

 
col_size = comb_size(tn_comb_mat)
row_size = set_size(tn_comb_mat)

ht = UpSet(tn_comb_mat, 
      row_title = "Species#SeqClassification", column_title = "Transposon Intersection across species",
      pt_size = unit(6, "pt"),
      lwd = unit(2, "pt"),
      #comb_col = "red",
      # bg_col = c("#d6b23b","#81dce2","#acf49a","#a784d8","#acf49a","#81dce2","#1a51dd","#1d7a01","#683cd8","#d62c4c","#683cd8","#1d7a01" , "#dd71c0","#1a51dd","#dd71c0","#d62c4c"),
      #bg_col = upset_row_col$Color,
      top_annotation = upset_top_annotation(tn_comb_mat, gp = gpar(fill = "#009797"), add_numbers = FALSE, 
                                            bar_width = 0.5, annotation_name_rot = 90),
      right_annotation = upset_right_annotation(tn_comb_mat, add_numbers = TRUE, gp = gpar(fill = "#009797")),
      show_row_names = TRUE,
      row_names_gp = gpar(fontsize = 9), #changes font size of "set size" labels
      width = unit(700, units = "pt"), height = unit(10, "cm"))

ht = draw(ht)
#ht
col_od = column_order(ht)
row_od = row_order(ht)

decorate_annotation("intersection_size", {
  grid.text(col_size[col_od], 
            seq_len(length(col_size)), 
            unit(col_size[col_od], "native") + unit(2, "mm"), 
            default.units = "native", just = "bottom",
            gp = gpar(fontsize = 8))
})
```

### Table4

#### Transposons proportions in Chromsome and plasmids by Species
```{r geneSharing_Transposon_props, fig.width=18, fig.height=10, echo=TRUE,message=FALSE}

bacant_TransposonCounts <- bacant_plasClass_df %>% 
  separate(Species_ContigClass, sep = "#", into = c("Species", "ContigClass")) %>%
  select(Species,Transposon,ContigClass) %>% 
  unique() %>% 
  mutate(Species_Transposon=paste(Species, Transposon, sep = '#')) %>% 
  select(Species_Transposon,ContigClass) %>% 
  table() %>% as.data.frame.matrix() %>% 
  mutate(BothChr_AND_Plasmid = case_when(Chromosome == 1 & Plasmid == 1 ~ 1, TRUE ~ 0)) %>% 
  tibble::rownames_to_column("Species#Transposon") %>% 
  separate("Species#Transposon", sep = "#", into = c("Species", "Transposon")) %>% 
  group_by(Species) %>% 
  summarise(Count_Transposon_in_Chr_old=sum(Chromosome), Count_Transposon_in_Plasmid_old=sum(Plasmid), Count_Transposon_in_Chr_Plasmid=sum(BothChr_AND_Plasmid)) %>% 
  mutate(Count_Transposon_in_Chr = Count_Transposon_in_Chr_old - Count_Transposon_in_Chr_Plasmid, 
         Count_Transposon_in_Plasmid = Count_Transposon_in_Plasmid_old - Count_Transposon_in_Chr_Plasmid) %>% 
  select(Species,Count_Transposon_in_Chr,Count_Transposon_in_Plasmid,Count_Transposon_in_Chr_Plasmid) %>% 
  column_to_rownames(var="Species")

# divides each cell value with corresponding row sum value
bacant_TransposonCounts_props <- bacant_TransposonCounts*100/rowSums(bacant_TransposonCounts) 

bacant_TransposonCounts_props %>% 
  kbl(caption = "Transposons proportions in Chromsome and plasmids by Species") %>% 
  kable_classic(full_width = F, html_font = "Cambria")
```

### Plot2

#### Transposons Percentages in Chromsome and plasmids by Species
```{r geneSharing_Transposon_props_plots, fig.width=18, fig.height=10, echo=TRUE,message=FALSE}
bacant_TransposonCounts_props %>% 
  rownames_to_column(var="Species") %>% 
  reshape2::melt(variable = "Transposon location", value.name = "Percentage") %>% 
  na_if(0) %>% 
  ggplot(aes(x = Species, y=Percentage, fill = `Transposon location`, label = Percentage)) + 
  geom_bar(width = 0.4, position = position_stack(), stat = "identity")+
  scale_fill_manual(labels=c('Chromosome', 'Plasmid',"Both"), values = c("#fcb6b1", "#facc96", "#4dbfbc")) +
  geom_text(aes(label=paste0(sprintf("%1.1f", Percentage),"%")), position=position_stack(vjust=0.5)) +
  scale_x_discrete(expand = c(0,0)) +
  ylab('Percentage')+
  xlab('Species')+
  ggtitle('Percentages of Transposons by their location')+
  theme_classic() +
  theme(axis.text.y=element_text(size=16,  vjust = 0, hjust = 1),
        axis.text.x=element_text(size=16, angle=90, vjust = 0, hjust = 1.2),
        axis.line=element_blank(),
        axis.ticks=element_blank(),
        legend.position="top") + coord_flip()
```


## Integrons {.tabset}

### Table1

#### Percentages of Samples carrying Integrons in Chromsomes and Plasmids across Species

```{r geneSharing_integron_samplepercentage, fig.width=18, fig.height=10, echo=TRUE,message=FALSE}

###@@@@@@@@@@@@@@ Dataframe2 - Bacant Predictions Integrons

bacAnt_Integrons_df <- data.table::fread(file = "2_bacant_annot/CPE_Trans_BacAnt_Integrons.tab", sep = "\t", header = TRUE)
#head(bacAnt_Integrons_df)

colnames(bacAnt_Integrons_df) <- c("Fasta", "Contig", "Fasta_Contig", "Integron")

#----------------Step2: Combining plasClass dataframe with bacant dataframe for plotting Transposons

bacAnt_Integrons_plasClass_df <- left_join(bacAnt_Integrons_df, plasClass_Prob_df, by = c("Fasta_Contig")) %>% 
  select(Fasta.x,Contig.x,Fasta_Contig,Integron,Classification,Species_ContigClass)

#head(bacAnt_Integrons_plasClass_df)

bacAnt_Integrons_plasClass_df <- bacAnt_Integrons_plasClass_df %>% filter(Fasta.x %in% hybridAssemb_1198list) # bacAnt_Integrons_plasClass_df now has only 1198 assembly information

# bacAnt_Integrons_plasClass_df %>%
#   select(Integron,Species_ContigClass) %>%
#   table() %>% # frequency of unique Integrons
#   kbl(caption = "Integrons and their counts in Chromsomes and Plasmids across Species") %>%
#   kable_classic(full_width = F, html_font = "Cambria")

bacAnt_Integrons_plasClass_df_total_col = bacAnt_Integrons_plasClass_df %>%
  select(Fasta.x,Integron,Species_ContigClass) %>%
  unique() %>%
  select(Species_ContigClass) %>% 
  unique() %>% 
  count() %>% c()

# percentages of Integrons
bacAnt_Integrons_plasClass_df %>% 
  select(Fasta.x,Integron,Species_ContigClass) %>%
  unique() %>% 
  select(Integron,Species_ContigClass) %>% 
  count(Integron, Species_ContigClass) %>% 
  rename(Freq = n) %>%  
  drop_na(Species_ContigClass) %>% 
  mutate(Species_ContigClass2=Species_ContigClass) %>% 
  separate(Species_ContigClass2, sep = "#", into = c("Species", "ContigClass")) %>% 
  left_join(Species_Counts, by = c("Species")) %>% 
  mutate(SamplePrcnt_wGeneElement=Freq*100/SampleCount) %>% 
  select(Integron,Species_ContigClass,SamplePrcnt_wGeneElement) %>% 
  data.table::dcast(Integron~Species_ContigClass, value.var="SamplePrcnt_wGeneElement") %>% 
  replace(is.na(.), 0) %>% 
  mutate_if(is.numeric, round, digits=1) %>% 
  #kbl(caption = "Sample count percentages for the Integrons in Chromsomes and Plasmids across Species") %>% 
  #kable_classic(full_width = F, html_font = "Cambria")
  formattable(., align = c("l",rep("c", bacAnt_Integrons_plasClass_df_total_col$n)), 
              list(
    `Transposon` = formatter("span", style = ~ style(color = "grey",font.weight = "bold")), 
    area(col = 1:bacAnt_Integrons_plasClass_df_total_col$n+1) ~ color_tile(customGreen0, customRed))) %>% 
  as.datatable(escape = FALSE,
               options = list(scrollX = FALSE, dom = 'lfti',
                              lengthMenu = list(c(5, 10, 25, 100, -1), c('5','10','25', '100', 'All')),
                              pageLength = 5),
               rownames = FALSE)
```


### Table2

```{r geneSharing_integron_count, fig.width=18, fig.height=10, echo=TRUE,message=FALSE}
bacAnt_Integrons_plasClass_df %>% 
  select(Integron,Species_ContigClass) %>% 
  unique() %>% 
  table() %>% # frequency of unique Integrons
  kbl(caption = "Integrons and their unique counts in Chromsomes and Plasmids across Species") %>% 
  kable_classic(full_width = F, html_font = "Cambria")
```

### Integrons UpsetR 

```{r geneSharing_integron_UpsetR, fig.width=18, fig.height=10, echo=TRUE,message=FALSE}

#----------------Step3: UpsetR Plot Integrons

integron_upsetR_df <- bacAnt_Integrons_plasClass_df %>% select(Integron,Species_ContigClass)
#View(integron_upsetR_df)
integron_upsetR_lt <- split(integron_upsetR_df$Integron, integron_upsetR_df$Species_ContigClass) 
integron_comb_mat = make_comb_mat(integron_upsetR_lt)

#integron_comb_mat
#str(integron_comb_mat)

#as.data.frame(integron_comb_mat)
# comb_mat
#UpSet(integron_comb_mat)


int_col_size = comb_size(integron_comb_mat)
int_row_size = set_size(integron_comb_mat)

integron_ht = UpSet(integron_comb_mat, 
           row_title = "Species#SeqClassification", column_title = "Integron Intersection across species",
           pt_size = unit(6, "pt"),
           lwd = unit(2, "pt"),
           #comb_col = "red",
#           bg_col = c("#d6b23b","#81dce2","#acf49a","#a784d8","#acf49a","#81dce2","#1a51dd","#1d7a01","#683cd8","#d62c4c","#683cd8","#1d7a01" , "#dd71c0","#1a51dd","#dd71c0","#d62c4c"),
           #bg_col = upset_row_col$Color,
           top_annotation = upset_top_annotation(integron_comb_mat, gp = gpar(fill = "#009797"), add_numbers = FALSE, 
                                                 bar_width = 0.5, annotation_name_rot = 90),
           right_annotation = upset_right_annotation(integron_comb_mat, add_numbers = TRUE, gp = gpar(fill = "#009797")),
           show_row_names = TRUE,
           row_names_gp = gpar(fontsize = 9), #changes font size of "set size" labels
           width = unit(700, units = "pt"), height = unit(10, "cm"))

ht = draw(integron_ht)
#ht
int_col_od = column_order(integron_ht)
int_row_od = row_order(integron_ht)

decorate_annotation("intersection_size", {
  grid.text(int_col_size[int_col_od], 
            seq_len(length(int_col_size)), 
            unit(int_col_size[int_col_od], "native") + unit(2, "mm"), 
            default.units = "native", just = "bottom",
            gp = gpar(fontsize = 8))
})
```

### Table4

#### Integrons proportions in Chromsome and plasmids by Species
```{r geneSharing_Integron_props, fig.width=18, fig.height=10, echo=TRUE,message=FALSE}

bacant_IntegronCounts <- bacAnt_Integrons_plasClass_df %>% 
  separate(Species_ContigClass, sep = "#", into = c("Species", "ContigClass")) %>%
  select(Species,Integron,ContigClass) %>% 
  unique() %>% 
  mutate(Species_Integron=paste(Species, Integron, sep = '#')) %>% 
  select(Species_Integron,ContigClass) %>% 
  table() %>% as.data.frame.matrix() %>% 
  mutate(BothChr_AND_Plasmid = case_when(Chromosome == 1 & Plasmid == 1 ~ 1, TRUE ~ 0)) %>% 
  tibble::rownames_to_column("Species#Integron") %>% 
  separate("Species#Integron", sep = "#", into = c("Species", "Integron")) %>% 
  group_by(Species) %>% 
  summarise(Count_Integron_in_Chr_old=sum(Chromosome), Count_Integron_in_Plasmid_old=sum(Plasmid), Count_Integron_in_Chr_Plasmid=sum(BothChr_AND_Plasmid)) %>% 
  mutate(Count_Integron_in_Chr = Count_Integron_in_Chr_old - Count_Integron_in_Chr_Plasmid, 
         Count_Integron_in_Plasmid = Count_Integron_in_Plasmid_old - Count_Integron_in_Chr_Plasmid) %>% 
  select(Species,Count_Integron_in_Chr,Count_Integron_in_Plasmid,Count_Integron_in_Chr_Plasmid) %>% 
  column_to_rownames(var="Species")

# divides each cell value with corresponding row sum value
bacant_IntegronCounts_props <- bacant_IntegronCounts*100/rowSums(bacant_IntegronCounts) 

bacant_IntegronCounts_props %>% 
  kbl(caption = "Integrons proportions in Chromsome and plasmids by Species") %>% 
  kable_classic(full_width = F, html_font = "Cambria")
```

### Plot2

#### Integrons Percentages in Chromsome and plasmids by Species
```{r geneSharing_Integron_props_plots, fig.width=18, fig.height=10, echo=TRUE,message=FALSE}
bacant_IntegronCounts_props %>% 
  rownames_to_column(var="Species") %>% 
  reshape2::melt(variable = "Integron location", value.name = "Percentage") %>% 
  na_if(0) %>% 
  ggplot(aes(x = Species, y=Percentage, fill = `Integron location`, label = Percentage)) + 
  geom_bar(width = 0.4, position = position_stack(), stat = "identity")+
  scale_fill_manual(labels=c('Chromosome', 'Plasmid',"Both"), values = c("#fcb6b1", "#facc96", "#4dbfbc")) +
  geom_text(aes(label=paste0(sprintf("%1.1f", Percentage),"%")), position=position_stack(vjust=0.5)) +
  scale_x_discrete(expand = c(0,0)) +
  ylab('Percentage')+
  xlab('Species')+
  ggtitle('Percentages of Integrons by their location')+
  theme_classic() +
  theme(axis.text.y=element_text(size=16,  vjust = 0, hjust = 1),
        axis.text.x=element_text(size=16, angle=90, vjust = 0, hjust = 1.2),
        axis.line=element_blank(),
        axis.ticks=element_blank(),
        legend.position="top") + coord_flip()
```

## AMR Genes {.tabset}

### Table1

#### Percentages of Samples carrying AMR genes in Chromsomes and Plasmids across Species
```{r geneSharing_AMRgenes_samplepercentage, echo=TRUE, fig.height=10, fig.width=18, message=FALSE}
###@@@@@@@@@@@@@@ Dataframe3 - AMRfinderplus - AMR Genes

amrFinderplus_df <- data.table::fread(file = "3_amrfinderplus_results/CPE_Trans_AMRFinderplus_AMR_Genes.tab", sep = "\t", header = TRUE)
#head(amrFinderplus_df)

colnames(amrFinderplus_df) <- c("Fasta", "Contig", "AMR_GENE", "AMR_GENE_CLASS","Coverage","Identity")

amrFinderplus_df <- amrFinderplus_df %>% filter(Fasta %in% hybridAssemb_1198list) # bacant_plasClass_df now has only 1198 assembly information

amrFinderplus_df <- 
  amrFinderplus_df %>%  
  mutate(Fasta_Contig=paste(Fasta, Contig, sep = '#'))

#----------------Step2: Combining plasClass dataframe with AMRFinderplus dataframe for plotting AMR genes

amrFinderplus_df <- left_join(amrFinderplus_df, plasClass_Prob_df, by = c("Fasta_Contig")) %>% 
  select(Fasta.x,Contig.x,AMR_GENE,AMR_GENE_CLASS,Classification,Species_ContigClass)

amrFinderplus_df_total_col = amrFinderplus_df %>%
  select(Fasta.x,AMR_GENE,Species_ContigClass) %>%
  unique() %>%
  select(Species_ContigClass) %>% 
  unique() %>% 
  count() %>% c()

# percentages of AMR_GENEs
amrFinderplus_df %>% 
  select(Fasta.x,AMR_GENE,Species_ContigClass) %>%
  unique() %>% 
  select(AMR_GENE,Species_ContigClass) %>% 
  count(AMR_GENE, Species_ContigClass) %>% 
  rename(Freq = n) %>%  
  drop_na(Species_ContigClass) %>% 
  mutate(Species_ContigClass2=Species_ContigClass) %>% 
  separate(Species_ContigClass2, sep = "#", into = c("Species", "ContigClass")) %>% 
  left_join(Species_Counts, by = c("Species")) %>% 
  mutate(SamplePrcnt_wGeneElement=Freq*100/SampleCount) %>% 
  select(AMR_GENE,Species_ContigClass,SamplePrcnt_wGeneElement) %>% 
  data.table::dcast(AMR_GENE~Species_ContigClass, value.var="SamplePrcnt_wGeneElement") %>% 
  replace(is.na(.), 0) %>% 
  mutate_if(is.numeric, round, digits=1) %>% 
  #kbl(caption = "Sample count percentages for the AMR genes in Chromsomes and Plasmids across Species") %>% 
  #kable_classic(full_width = F, html_font = "Cambria")
  formattable(., align = c("l",rep("c", amrFinderplus_df_total_col$n)), 
              list(
    `Transposon` = formatter("span", style = ~ style(color = "grey",font.weight = "bold")), 
    area(col = 1:amrFinderplus_df_total_col$n+1) ~ color_tile(customGreen0, customRed))) %>% 
  as.datatable(escape = FALSE,
               options = list(scrollX = FALSE, dom = 'lfti',
                              lengthMenu = list(c(5,10,25, 100, -1), c('5','10','25', '100', 'All')),
                              pageLength = 5),
               rownames = FALSE)
```

### Table2

```{r geneSharing_AMRgenes_count, fig.width=18, fig.height=10, echo=TRUE,message=FALSE}
### AMR Genes and their unique counts in Chromsomes and Plasmids across Species

amrFinderplus_df %>% 
  select(AMR_GENE,Species_ContigClass) %>% 
  unique() %>% 
  table() %>% # frequency of unique transposons
  kbl(caption = "AMR Genes and their unique counts in Chromsomes and Plasmids across Species") %>% 
  kable_classic(full_width = F, html_font = "Cambria")
```

### AMR Genes UpsetR 

```{r geneSharing_AMRgenes_UpsetR, fig.width=18, fig.height=10, echo=TRUE,message=FALSE}
#----------------Step3: UpsetR Plot AMR Genes
amr_upsetR_df <- amrFinderplus_df %>% select(AMR_GENE,Species_ContigClass)

amr_upsetR_lt <- split(amr_upsetR_df$AMR_GENE, amr_upsetR_df$Species_ContigClass) 
amr_comb_mat = make_comb_mat(amr_upsetR_lt)

col_size = comb_size(amr_comb_mat)
row_size = set_size(amr_comb_mat)

amr_ht = UpSet(amr_comb_mat, 
           row_title = "Species#SeqClassification", column_title = "AMR Genes Intersection across species",
           pt_size = unit(6, "pt"),
           lwd = unit(2, "pt"),
           #comb_col = "red",
           # bg_col = c("#d6b23b","#81dce2","#acf49a","#a784d8","#acf49a","#81dce2","#1a51dd","#1d7a01","#683cd8","#d62c4c","#683cd8","#1d7a01" ,
           #            "#dd71c0","#1a51dd","#dd71c0","#d62c4c"),
           #bg_col = upset_row_col$Color,
           top_annotation = upset_top_annotation(amr_comb_mat, gp = gpar(fill = "#009797"), add_numbers = FALSE, 
                                                 bar_width = 0.5, annotation_name_rot = 90),
           right_annotation = upset_right_annotation(amr_comb_mat, add_numbers = TRUE, gp = gpar(fill = "#009797")),
           show_row_names = TRUE,
           row_names_gp = gpar(fontsize = 9), #changes font size of "set size" labels
           width = unit(900, units = "pt"), height = unit(12, "cm"))

amr_ht = draw(amr_ht)
#ht
col_od = column_order(amr_ht)
row_od = row_order(amr_ht)

decorate_annotation("intersection_size", {
  grid.text(col_size[col_od], 
            seq_len(length(col_size)), 
            unit(col_size[col_od], "native") + unit(2, "mm"), 
            default.units = "native", just = "bottom",
            gp = gpar(fontsize = 8))
})
```

### Table3

#### AMR Genes counts in Chromsome and plasmids by Species
```{r geneSharing_AMRgenes_counts, fig.width=18, fig.height=10, echo=TRUE,message=FALSE}

amrFinderplus_df %>% 
  separate(Species_ContigClass, sep = "#", into = c("Species", "ContigClass")) %>% 
  select(Species,AMR_GENE,ContigClass) %>% 
  unique() %>% 
  mutate(Species_AMR_GENE=paste(Species, AMR_GENE, sep = '#')) %>% 
  select(Species_AMR_GENE,ContigClass) %>% 
  table() %>% as.data.frame.matrix() %>% 
  mutate(BothChr_AND_Plasmid = case_when(Chromosome == 1 & Plasmid == 1 ~ 1, TRUE ~ 0)) %>% 
  tibble::rownames_to_column("Species#AMRGEnes") %>% 
  separate("Species#AMRGEnes", sep = "#", into = c("Species", "AMRGene")) %>% 
  group_by(Species) %>% 
  summarise(Count_AMRgenes_in_Chr_old=sum(Chromosome), Count_AMRgenes_in_Plasmid_old=sum(Plasmid), Count_AMRgenes_in_Chr_Plasmid=sum(BothChr_AND_Plasmid)) %>% 
  mutate(Count_AMRgenes_in_Chr = Count_AMRgenes_in_Chr_old - Count_AMRgenes_in_Chr_Plasmid, 
         Count_AMRgenes_in_Plasmid = Count_AMRgenes_in_Plasmid_old - Count_AMRgenes_in_Chr_Plasmid) %>% 
  select(Species,Count_AMRgenes_in_Chr,Count_AMRgenes_in_Plasmid,Count_AMRgenes_in_Chr_Plasmid) %>% 
  mutate(Total_AMRGenes = rowSums(.[2:4])) %>% 
  kbl(caption = "AMR Genes counts in Chromsome and plasmids by Species") %>% 
  kable_classic(full_width = F, html_font = "Cambria")

```

### Table4

#### AMR Genes proportions in Chromsome and plasmids by Species
```{r geneSharing_AMRgenes_props, fig.width=18, fig.height=10, echo=TRUE,message=FALSE}

amr_geneCounts <- amrFinderplus_df %>% 
  separate(Species_ContigClass, sep = "#", into = c("Species", "ContigClass")) %>% 
  select(Species,AMR_GENE,ContigClass) %>% 
  unique() %>% 
  mutate(Species_AMR_GENE=paste(Species, AMR_GENE, sep = '#')) %>% 
  select(Species_AMR_GENE,ContigClass) %>% 
  table() %>% as.data.frame.matrix() %>% 
  mutate(BothChr_AND_Plasmid = case_when(Chromosome == 1 & Plasmid == 1 ~ 1, TRUE ~ 0)) %>% 
  tibble::rownames_to_column("Species#AMRGEnes") %>% 
  separate("Species#AMRGEnes", sep = "#", into = c("Species", "AMRGene")) %>% 
  group_by(Species) %>% 
  summarise(Count_AMRgenes_in_Chr_old=sum(Chromosome), Count_AMRgenes_in_Plasmid_old=sum(Plasmid), Count_AMRgenes_in_Chr_Plasmid=sum(BothChr_AND_Plasmid)) %>% 
  mutate(Count_AMRgenes_in_Chr = Count_AMRgenes_in_Chr_old - Count_AMRgenes_in_Chr_Plasmid, 
         Count_AMRgenes_in_Plasmid = Count_AMRgenes_in_Plasmid_old - Count_AMRgenes_in_Chr_Plasmid) %>% 
  select(Species,Count_AMRgenes_in_Chr,Count_AMRgenes_in_Plasmid,Count_AMRgenes_in_Chr_Plasmid) %>% 
  column_to_rownames(var="Species")

# divides each cell value with corresponding row sum value
amr_geneCounts_props <- amr_geneCounts*100/rowSums(amr_geneCounts) 

amr_geneCounts_props %>% 
  kbl(caption = "AMR Genes proportions in Chromsome and plasmids by Species") %>% 
  kable_classic(full_width = F, html_font = "Cambria")
```

### Plot2

#### AMR Genes Percentages in Chromsome and plasmids by Species
```{r geneSharing_AMRgenes_props_plots, fig.width=18, fig.height=10, echo=TRUE,message=FALSE}
amr_geneCounts_props %>% 
  rownames_to_column(var="Species") %>% 
  reshape2::melt(variable = "AMR gene location", value.name = "Percentage") %>% 
  na_if(0) %>% 
  ggplot(aes(x = Species, y=Percentage, fill = `AMR gene location`, label = Percentage)) + 
  geom_bar(width = 0.4, position = position_stack(), stat = "identity")+
  scale_fill_manual(labels=c('Chromosome', 'Plasmid',"Both"), values = c("#fcb6b1", "#facc96", "#4dbfbc")) +
  geom_text(aes(label=paste0(sprintf("%1.1f", Percentage),"%")), position=position_stack(vjust=0.5)) +
  scale_x_discrete(expand = c(0,0)) +
  ylab('Percentage')+
  xlab('Species')+
  ggtitle('Percentages of AMR genes by their location')+
  theme_classic() +
  theme(axis.text.y=element_text(size=16,  vjust = 0, hjust = 1),
        axis.text.x=element_text(size=16, angle=90, vjust = 0, hjust = 1.2),
        axis.line=element_blank(),
        axis.ticks=element_blank(),
        legend.position="top") + coord_flip()
```

## VF Genes {.tabset}

### Table1

#### Percentages of Samples carrying Virulence genes in Chromsomes and Plasmids across Species
```{r geneSharing_VFgenes_samplepercentage, fig.width=18, fig.height=10, echo=TRUE,message=FALSE}


###@@@@@@@@@@@@@@ Dataframe4 - Abricate - Virulence Factor Genes

Abricate_vf_df <- data.table::fread(file = "4_abricate_vf_results/CPE_Trans_Abricate_VF_Genes.tab", sep = "\t", header = TRUE)
#head(Abricate_vf_df)

##########-------------- Modifying dataframe START -------------------################

colnames(Abricate_vf_df ) <- c("Fasta", "Contig","START","END","STRAND","VF_GENE")

Abricate_vf_df <- 
  Abricate_vf_df %>%  
  mutate(Fasta_Contig=paste(Fasta, Contig, sep = '#'))

Abricate_vf_df <- Abricate_vf_df %>% filter(Fasta %in% hybridAssemb_1198list) # bacant_plasClass_df now has only 1198 assembly information

#head(Abricate_vf_df)

#----------------Step2: Combining plasClass dataframe with abricate_vf_dataframe for plotting virulence genes

Abricate_vf_df <- left_join(Abricate_vf_df, plasClass_Prob_df, by = c("Fasta_Contig")) %>% 
  select(Fasta.x,Contig.x,VF_GENE,Classification,Species_ContigClass)

#View(Abricate_vf_df)

Abricate_vf_df_total_col = Abricate_vf_df %>%
  select(Fasta.x,VF_GENE,Species_ContigClass) %>%
  unique() %>%
  select(Species_ContigClass) %>% 
  unique() %>% 
  drop_na(Species_ContigClass) %>% count() %>% c()

# percentages of AMR_GENEs
Abricate_vf_df %>% 
  select(Fasta.x,VF_GENE,Species_ContigClass) %>%
  unique() %>% 
  select(VF_GENE,Species_ContigClass) %>% 
  count(VF_GENE, Species_ContigClass) %>% 
  rename(Freq = n) %>%  
  drop_na(Species_ContigClass) %>% 
  mutate(Species_ContigClass2=Species_ContigClass) %>% 
  separate(Species_ContigClass2, sep = "#", into = c("Species", "ContigClass")) %>% 
  left_join(Species_Counts, by = c("Species")) %>% 
  mutate(SamplePrcnt_wGeneElement=Freq*100/SampleCount) %>% 
  select(VF_GENE,Species_ContigClass,SamplePrcnt_wGeneElement) %>% 
  data.table::dcast(VF_GENE~Species_ContigClass, value.var="SamplePrcnt_wGeneElement") %>% 
  replace(is.na(.), 0) %>% 
  mutate_if(is.numeric, round, digits=1) %>% 
 # kbl(caption = "Sample count percentages for the Virulence factor genes in Chromsomes and Plasmids across Species") %>% 
 # kable_classic(full_width = F, html_font = "Cambria")
  formattable(., align = c("l",rep("c", Abricate_vf_df_total_col $n)), 
              list(
    `Transposon` = formatter("span", style = ~ style(color = "grey",font.weight = "bold")), 
    area(col = 1:Abricate_vf_df_total_col $n+1) ~ color_tile(customGreen0, customRed))) %>% 
  as.datatable(escape = FALSE,
               options = list(scrollX = FALSE, dom = 'lfti',
                              lengthMenu = list(c(5, 10, 25, 100, -1), c('5','10','25', '100', 'All')),
                              pageLength = 5),
               rownames = FALSE)
```

### Table2

#### Virulence genes and their unique counts in Chromsomes and Plasmids across Species
```{r geneSharing_vfgenes_count, fig.width=18, fig.height=10, echo=TRUE,message=FALSE}

Abricate_vf_df %>% 
  select(VF_GENE,Species_ContigClass) %>% 
  unique() %>% 
  table() %>% # frequency of unique transposons
  kbl(caption = "Virulence factors genes and their unique counts in Chromsomes and Plasmids across Species") %>% 
  kable_classic(full_width = F, html_font = "Cambria")
```

### VF Genes UpsetR 

```{r geneSharing_VFgenes_UpsetR, fig.width=18, fig.height=10, echo=TRUE,message=FALSE}

#----------------Step3: UpsetR Plot VFGenes

abr_upsetR_df <- Abricate_vf_df %>% select(VF_GENE,Species_ContigClass)
#View(abr_upsetR_df)

abr_upsetR_lt <- split(abr_upsetR_df$VF_GENE, abr_upsetR_df$Species_ContigClass) 
abr_comb_mat = make_comb_mat(abr_upsetR_lt)

col_size = comb_size(abr_comb_mat)
row_size = set_size(abr_comb_mat)

vf_ht = UpSet(abr_comb_mat, 
           row_title = "Species#SeqClassification", column_title = "Virulence factor genes Intersection across species",
           pt_size = unit(6, "pt"),
           lwd = unit(2, "pt"),
           #comb_col = "red",
           # bg_col = c("#d6b23b","#81dce2","#acf49a","#a784d8","#acf49a","#81dce2","#1a51dd","#1d7a01","#683cd8","#d62c4c","#683cd8","#1d7a01" ,
           #            "#dd71c0","#1a51dd","#dd71c0","#d62c4c"),
           #bg_col = upset_row_col$Color,
           top_annotation = upset_top_annotation(abr_comb_mat, gp = gpar(fill = "#009797"), add_numbers = FALSE, 
                                                 bar_width = 0.5, annotation_name_rot = 90),
           right_annotation = upset_right_annotation(abr_comb_mat, add_numbers = TRUE, gp = gpar(fill = "#009797")),
           show_row_names = TRUE,
           row_names_gp = gpar(fontsize = 9), #changes font size of "set size" labels
           width = unit(900, units = "pt"), height = unit(12, "cm"))

vf_ht = draw(vf_ht)
#ht
col_od = column_order(vf_ht)
row_od = row_order(vf_ht)

decorate_annotation("intersection_size", {
  grid.text(col_size[col_od], 
            seq_len(length(col_size)), 
            unit(col_size[col_od], "native") + unit(2, "mm"), 
            default.units = "native", just = "bottom",
            gp = gpar(fontsize = 8))
})
```

### Table4

#### VF Genes proportions in Chromsome and plasmids by Species
```{r geneSharing_VFgenes_props, fig.width=18, fig.height=10, echo=TRUE,message=FALSE}
vf_geneCounts <- Abricate_vf_df %>% 
  separate(Species_ContigClass, sep = "#", into = c("Species", "ContigClass")) %>%
  select(Species,VF_GENE,ContigClass) %>% 
  unique() %>% 
  mutate(Species_VF_GENE=paste(Species, VF_GENE, sep = '#')) %>% 
  select(Species_VF_GENE,ContigClass) %>% 
  table() %>% as.data.frame.matrix() %>% 
  mutate(BothChr_AND_Plasmid = case_when(Chromosome == 1 & Plasmid == 1 ~ 1, TRUE ~ 0)) %>% 
  tibble::rownames_to_column("Species#VFgenes") %>% 
  separate("Species#VFgenes", sep = "#", into = c("Species", "VFGene")) %>% 
  group_by(Species) %>% 
  summarise(Count_VFgenes_in_Chr_old=sum(Chromosome), Count_VFgenes_in_Plasmid_old=sum(Plasmid), Count_VFgenes_in_Chr_Plasmid=sum(BothChr_AND_Plasmid)) %>% 
  mutate(Count_VFgenes_in_Chr = Count_VFgenes_in_Chr_old - Count_VFgenes_in_Chr_Plasmid, 
         Count_VFgenes_in_Plasmid = Count_VFgenes_in_Plasmid_old - Count_VFgenes_in_Chr_Plasmid) %>% 
  select(Species,Count_VFgenes_in_Chr,Count_VFgenes_in_Plasmid,Count_VFgenes_in_Chr_Plasmid) %>% 
  column_to_rownames(var="Species")

# divides each cell value with corresponding row sum value
vf_geneCounts_props <- vf_geneCounts*100/rowSums(vf_geneCounts) 

vf_geneCounts_props %>% 
  kbl(caption = "VF Genes proportions in Chromsome and plasmids by Species") %>% 
  kable_classic(full_width = F, html_font = "Cambria")
```

### Plot2

#### VF Genes Percentages in Chromsome and plasmids by Species
```{r geneSharing_VFgenes_props_plots, fig.width=18, fig.height=10, echo=TRUE,message=FALSE}
vf_geneCounts_props %>% 
  rownames_to_column(var="Species") %>% 
  reshape2::melt(variable = "VF Gene location", value.name = "Percentage") %>% 
  na_if(0) %>% 
  ggplot(aes(x = Species, y=Percentage, fill = `VF Gene location`, label = Percentage)) + 
  geom_bar(width = 0.4, position = position_stack(), stat = "identity")+
  scale_fill_manual(labels=c('Chromosome', 'Plasmid',"Both"), values = c("#fcb6b1", "#facc96", "#4dbfbc")) +
  geom_text(aes(label=paste0(sprintf("%1.1f", Percentage),"%")), position=position_stack(vjust=0.5)) +
  scale_x_discrete(expand = c(0,0)) +
  ylab('Percentage')+
  xlab('Species')+
  ggtitle('Percentages of Virulence factor genes by their location')+
  theme_classic() +
  theme(axis.text.y=element_text(size=16,  vjust = 0, hjust = 1),
        axis.text.x=element_text(size=16, angle=90, vjust = 0, hjust = 1.2),
        axis.line=element_blank(),
        axis.ticks=element_blank(),
        legend.position="top") + coord_flip()
```


## Insertion Sequence (IS) {.tabset}

### Table1

#### Percentages of Samples carrying Insertion Sequences in Chromsomes and Plasmids across Species

```{r geneSharing_IS_samplepercentage, fig.width=18, fig.height=10, echo=TRUE,message=FALSE}
###@@@@@@@@@@@@@@ Dataframe6 - Prokka Insertion Sequences 

Prokka_IS_df <- data.table::fread(file = "6_Prokka_InsertionSequences/CPE_Trans_Prokka_IS.tab", sep = "\t", header = TRUE)
#head(Prokka_IS_df)

colnames(Prokka_IS_df) <- c("Fasta", "Contig", "Fasta_Contig", "IS")

#----------------Step2: Combining plasClass dataframe with Prokka_IS dataframe for plotting Transposons

Prokka_IS_plasClass_df <- left_join(Prokka_IS_df, plasClass_Prob_df, by = c("Fasta_Contig")) %>% 
  select(Fasta.x,Contig.x,Fasta_Contig,IS,Classification,Species_ContigClass)

#head(Prokka_IS_plasClass_df)

Prokka_IS_plasClass_df <- Prokka_IS_plasClass_df %>% filter(Fasta.x %in% hybridAssemb_1198list) # Prokka_IS_plasClass_df now has only 1198 assembly information

# Prokka_IS_plasClass_df %>%
#   select(Transposon,Species_ContigClass) %>%
#   table() %>% # frequency of unique transposons
#   kbl(caption = "Transposons and their counts in Chromosomes and Plasmids across Species") %>%
#   kable_classic(full_width = F, html_font = "Cambria")

customGreen0 = "#DeF7E9"
customRed = "#ff7f7f"
Prokka_IS_plasClass_df_total_col = Prokka_IS_plasClass_df %>% 
  select(Fasta.x,IS,Species_ContigClass) %>% 
  unique() %>%
  select(Species_ContigClass) %>% 
  unique() %>% 
  count() %>% c()


# percentages of IS (https://clarewest.github.io/blog/post/making-tables-shiny/)
Prokka_IS_plasClass_df %>%
  select(Fasta.x,IS,Species_ContigClass) %>%
  unique() %>%
  select(IS,Species_ContigClass) %>%
  count(IS, Species_ContigClass) %>%
  rename(Freq = n) %>%
  drop_na(Species_ContigClass) %>%
  mutate(Species_ContigClass2=Species_ContigClass) %>%
  separate(Species_ContigClass2, sep = "#", into = c("Species", "ContigClass")) %>%
  left_join(Species_Counts, by = c("Species")) %>%
  mutate(SamplePrcnt_wGeneElement=Freq*100/SampleCount) %>%
  select(IS,Species_ContigClass,SamplePrcnt_wGeneElement) %>%
  data.table::dcast(IS~Species_ContigClass, value.var="SamplePrcnt_wGeneElement") %>%
  replace(is.na(.), 0) %>%
  mutate_if(is.numeric, round, digits=1) %>% 
  formattable(., align = c("l",rep("c", Prokka_IS_plasClass_df_total_col$n)), 
              list(
                `IS` = formatter("span", style = ~ style(color = "grey",font.weight = "bold")), 
                area(col = 1:Prokka_IS_plasClass_df_total_col$n+1) ~ color_tile(customGreen0, customRed))) %>% 
  as.datatable(escape = FALSE,
               options = list(scrollX = FALSE, dom = 'lfti',
                              lengthMenu = list(c(5, 10, 25, 100, -1), c('5','10','25', '100', 'All')),
                              pageLength = 5),
               rownames = FALSE)

```

### Table2

#### Insertion Sequences and their unique counts in Chromsomes and Plasmids across Species
```{r geneSharing_IS_count, fig.width=18, fig.height=10, echo=TRUE,message=FALSE}
Prokka_IS_plasClass_df %>% 
  select(IS,Species_ContigClass) %>% 
  unique() %>% 
  table() %>% # frequency of unique transposons
  kbl(caption = "Insertion Sequence and their unique counts in Chromsomes and Plasmids across Species") %>% 
  kable_classic(full_width = F, html_font = "Cambria")
```

### Insertion Sequences UpsetR 

```{r geneSharing_IS_UpsetR, fig.width=18, fig.height=10, echo=TRUE,message=FALSE}

#----------------Step3: UpsetR Plot VFGenes

IS_upsetR_df <- Prokka_IS_plasClass_df  %>% select(IS,Species_ContigClass)
#View(IS_upsetR_df)

IS_upsetR_lt <- split(IS_upsetR_df$IS, IS_upsetR_df$Species_ContigClass) 
IS_comb_mat = make_comb_mat(IS_upsetR_lt)

col_size = comb_size(IS_comb_mat)
row_size = set_size(IS_comb_mat)

IS_ht = UpSet(IS_comb_mat, 
           row_title = "Species#SeqClassification", column_title = "Insertion Sequences Intersection across species",
           pt_size = unit(6, "pt"),
           lwd = unit(2, "pt"),
           #comb_col = "red",
           # bg_col = c("#d6b23b","#81dce2","#acf49a","#a784d8","#acf49a","#81dce2","#1a51dd","#1d7a01","#683cd8","#d62c4c","#683cd8","#1d7a01" ,
           #            "#dd71c0","#1a51dd","#dd71c0","#d62c4c"),
           #bg_col = upset_row_col$Color,
           top_annotation = upset_top_annotation(IS_comb_mat, gp = gpar(fill = "#009797"), add_numbers = FALSE, 
                                                 bar_width = 0.5, annotation_name_rot = 90),
           right_annotation = upset_right_annotation(IS_comb_mat, add_numbers = TRUE, gp = gpar(fill = "#009797")),
           show_row_names = TRUE,
           row_names_gp = gpar(fontsize = 9), #changes font size of "set size" labels
           width = unit(900, units = "pt"), height = unit(12, "cm"))

IS_ht = draw(IS_ht)
#ht
col_od = column_order(IS_ht)
row_od = row_order(IS_ht)

decorate_annotation("intersection_size", {
  grid.text(col_size[col_od], 
            seq_len(length(col_size)), 
            unit(col_size[col_od], "native") + unit(2, "mm"), 
            default.units = "native", just = "bottom",
            gp = gpar(fontsize = 8))
})
```

### Table4

#### Insertion Sequences proportions in Chromsome and plasmids by Species
```{r geneSharing_IS_props, fig.width=18, fig.height=10, echo=TRUE,message=FALSE}

Prokka_ISCounts <- Prokka_IS_plasClass_df %>% 
  separate(Species_ContigClass, sep = "#", into = c("Species", "ContigClass")) %>%
  select(Species,IS,ContigClass) %>% 
  unique() %>% 
  mutate(Species_IS=paste(Species, IS, sep = '#')) %>% 
  select(Species_IS,ContigClass) %>% 
  table() %>% as.data.frame.matrix() %>% 
  mutate(BothChr_AND_Plasmid = case_when(Chromosome == 1 & Plasmid == 1 ~ 1, TRUE ~ 0)) %>% 
  tibble::rownames_to_column("Species#VFgenes") %>% 
  separate("Species#VFgenes", sep = "#", into = c("Species", "VFGene")) %>% 
  group_by(Species) %>% 
  summarise(Count_IS_in_Chr_old=sum(Chromosome), Count_IS_in_Plasmid_old=sum(Plasmid), Count_IS_in_Chr_Plasmid=sum(BothChr_AND_Plasmid)) %>% 
  mutate(Count_IS_in_Chr = Count_IS_in_Chr_old - Count_IS_in_Chr_Plasmid, 
         Count_IS_in_Plasmid = Count_IS_in_Plasmid_old - Count_IS_in_Chr_Plasmid) %>% 
  select(Species,Count_IS_in_Chr,Count_IS_in_Plasmid,Count_IS_in_Chr_Plasmid) %>% 
  column_to_rownames(var="Species")

# divides each cell value with corresponding row sum value
Prokka_ISCounts_props <- Prokka_ISCounts*100/rowSums(Prokka_ISCounts) 

Prokka_ISCounts_props %>% 
  kbl(caption = "Insertion Sequences proportions in Chromsome and plasmids by Species") %>% 
  kable_classic(full_width = F, html_font = "Cambria")
```

### Plot2

#### Insertion Sequences Percentages in Chromsome and plasmids by Species
```{r geneSharing_IS_props_plots, fig.width=18, fig.height=10, echo=TRUE,message=FALSE}
Prokka_ISCounts_props %>% 
  rownames_to_column(var="Species") %>% 
  reshape2::melt(variable = "Insertion Sequences location", value.name = "Percentage") %>% 
  na_if(0) %>% 
  ggplot(aes(x = Species, y=Percentage, fill = `Insertion Sequences location`, label = Percentage)) + 
  geom_bar(width = 0.4, position = position_stack(), stat = "identity")+
  scale_fill_manual(labels=c('Chromosome', 'Plasmid',"Both"), values = c("#fcb6b1", "#facc96", "#4dbfbc")) +
  geom_text(aes(label=paste0(sprintf("%1.1f", Percentage),"%")), position=position_stack(vjust=0.5)) +
  scale_x_discrete(expand = c(0,0)) +
  ylab('Percentage')+
  xlab('Species')+
  ggtitle('Percentages of Insertion Sequences by their location')+
  theme_classic() +
  theme(axis.text.y=element_text(size=16,  vjust = 0, hjust = 1),
        axis.text.x=element_text(size=16, angle=90, vjust = 0, hjust = 1.2),
        axis.line=element_blank(),
        axis.ticks=element_blank(),
        legend.position="top") + coord_flip()
```