---
title: "Data Analysis of Pseudomonas aeruginosa Samples"
author: "Prakki Sai Rama Sridatta"
date: "`r format(Sys.Date())`"
output:
  html_document:
    keep_md: true
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### Step 1: Raw read analysis fastqc and multiqc

The following analysis is only done 16 environmental samples. Rest 183 samples are previously done with analysis with the same steps.
The results of these 16 samples are appended to 183 samples. 

```{r rawdata, engine = 'bash', eval = FALSE}
#Location -
# cd /storage/data/DATA4/analysis/27_Paeruginosa_183_samples_Shawn_with_16_Envsamples

###################################################################

# Step 1.fastqc on Raw Reads 
##############################

mkdir 1_Rawdata

cd 1_Rawdata

# Copying softlinks of the raw data files
for d in $(cat Environmental_16_Samples.list); do ln -s $d .; done

# Moving into fastq files into respective folders
for d in $(ls *.gz | cut -f1 -d "_"  | sort -u); do mkdir $d; mv "$d"_*.gz $d; done

# Run fastqc on the raw reads (Env + 11F)
time for d in $(ls -d */| sed 's/\///g'); do echo $d; cd $d; /storage/apps/FastQC/fastqc -t 48 *.fq.gz; cd ..; done

#combine all fastqc reports using multiqc
cd 1_Rawdata

multiqc .

# Unzip the gz files - because these are softlinks we write to  an other file
cd 11F

time for x in $(ls -d */ | tr -d '/'); 
do 
cd $x; 
echo $x; 
  for d in $(ls *.fastq.gz); 
  do 
  base=`echo $d | sed 's/.fastq.gz//'`; 
  gunzip -c $d >"$base".fastq & 
  done; 
cd ..;
done

# Observations from raw fastq: Adapters are present in many sequences. So, trimmed them using BBMAP.
# But base quality of all the samples looking high (above Q30).

# For generating read stats - Run this on both Env and 11F folders
$ time for d in $(ls */*fq); do echo $d; perl /storage/apps/SNP_Validation_Scripts/tools/Q20_Q30_Stats_wo_functions.pl $d & done 

# TO find the Q20_Q30 Stats faster, lets write the commands to shell script file
$ time for d in $(for e in $(ls -l | grep -v '>' | awk '{print $NF}' | grep 'C'); do ls -1 $e/*.fq; done); do echo "perl /storage/apps/SNP_Validation_Scripts/tools/Q20_Q30_Stats_wo_functions.pl $d"; done >parallel_script_Q20_Q30_stats.sh
time parallel < parallel_script_Q20_Q30_stats.sh &

$ pwd
#/storage/data/DATA4/analysis/30_Aqueous_Env_11F_data_analysis/1_Rawdata

$ cat */Q30_Q20_readstats.txt >rawData_Q30_Q20_readstats.txt

# Run qualStats.R script with the rawData_Q30_Q20_readstats.txt as an input. This should generate "_Pre-Trimmed_aggregated_stats.txt" file

$ pwd
# /storage/data/DATA4/analysis/30_Aqueous_Env_11F_data_analysis/2_Cleandata/illumina_bbmap_trimmed

# Run qualStats.R script with the rawData_Q30_Q20_readstats.txt as an input. This should generate "_Pre-Trimmed_aggregated_stats.txt" file

# Run qualStats.R script with the filtData_Q30_Q20_readstats.txt as an input. This should generate "_Post-Trimmed_aggregated_stats.txt" file

# Now combine both Raw and Filt Data using qualStats_aggregated.R script


```

### Step 2: Pre-process Illumina reads 

Now let us trim the raw data from illumina using BBDuk tool

```{r trimdata_Illumina, engine = 'bash', eval = FALSE}
time for d in $(ls -d */); 
do 
  echo $d; 
  subdir=`echo $d`; 
  cd $subdir; 
  R1=`ls *_1.fq | sed 's/.fq//g'`; R2=`ls *_2.fq | sed 's/.fq//g'`; 
  echo "$R1 $R2"; 
  /storage/apps/bbmap/bbduk.sh -Xmx6g in1=$R1.fq in2=$R2.fq out1=$R1\_bbmap_adaptertrimmed.fq out2=$R2\_bbmap_adaptertrimmed.fq ref=/storage/apps/bbmap/resources/adapters.fa ktrim=r k=23 mink=11 hdist=1 qtrim=rl trimq=30 minavgquality=30; 
  cd ..; 
done

# Manually copied the bbmap command log into adapter_trimming.log

mkdir ../2_CleanData

# move the adapter trimmed files into another directory
mv */*bbmap_adaptertrimmed.fq  /storage/data/DATA4/analysis/37_Ecloacae_Scope_Patient_Analysis/2_CleanData
cd ../2_CleanData/

# move to specific folder
for d in $(ls *.fq| awk -F_ '{print $1}' | sort -u); do echo $d; mkdir $d; mv "$d"_*.fq $d; done

# Run fastqc on the raw reads (Env + 11F)
time for d in $(ls -d */| sed 's/\///g'); do echo $d; cd $d; /storage/apps/FastQC/fastqc -t 48 *.fq; cd ..; done

#combine all fastqc reports using multiqc
cd 1_Rawdata

multiqc .

$ pwd
/storage/data/DATA4/analysis/33_CRO_Transmission


```

Plotting Raw Read and Clean Read Stats


```{r Illumina_readData_plots, fig.width=12, fig.height=8, message=FALSE}

library(tidyr)
library(dplyr)

library(ggplot2)
library(scales)

setwd("/data02/Analysis/Projects/8_Aqueos_samples/1_ReadData_Stats_Raw_vs_Filt")

stats <- read.csv("Aqueous_samples_AITBiotech_Pre_AND_Post_Trimming_Stats.csv", header = TRUE, sep = ",")
head(stats)

############--------CHANGE THESE VARIABLES---IMPORTANT!!------##########

samples_or_project_name <- "Aqueous"
seqCompany <- "AITBiotech" #Internal/AITBiotech
seqData_trimmed_or_not <- "Pre and Post-Trimmed" # use "Raw data" before trimming else use "After Trimming with Q30 score"  
genomeSize <- 5500000 


# Plotting

###----> Genome Coverage

ggplot(stats[order(pmax(stats$Genome_Coverage_Pre_Trimming, stats$Genome_Coverage_Post_Trimming)),], aes (x=factor(Isolate, levels=Isolate), Genome_Coverage)) +  
  geom_point(aes(y=Genome_Coverage_Pre_Trimming, color = "Genome_Coverage_Pre_Trimming"), size=1) + 
  geom_point(aes(y=Genome_Coverage_Post_Trimming, color = "Genome_Coverage_Post_Trimming"), size=1) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size=5),legend.position = "top") + #Rotating and spacing axis labels
  expand_limits(x = c(0, NA), y = c(0, NA)) +
  xlab("Isolates") +
  ylab("Genome Coverage ") +
  ggtitle(paste0("Genome Coverage of ",samples_or_project_name," samples from ",seqCompany," (",seqData_trimmed_or_not,")")) #+ 

#  facet_wrap(~Species,scales = "free_x", nrow = 2) # ggtitle("Genome Coverage of Steno Samples from AITBiotech (After Trimming)")

ggsave("GenomeCoverage.png",dpi = 600, width = 8, height = 6, units = "in")

###----> Total Reads

ggplot(stats[order(pmax(stats$TotalReads_Pre_Trimming, stats$TotalReads_Post_Trimming)),], aes (x=factor(Isolate, levels=Isolate), Total_Reads)) +  
  geom_point(aes (y= TotalReads_Pre_Trimming, color = "TotalReads_Pre_Trimming"), size=1) + 
  geom_point(aes(y=TotalReads_Post_Trimming, color = "TotalReads_Post_Trimming"), size=1) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size=5),legend.position = "top") + #Rotating and spacing axis labels
  expand_limits(x = c(0, NA), y = c(0, NA)) +
  scale_y_continuous(labels = unit_format(unit = "M", scale = 1e-6)) + ## need to load "scales" library - very readable
  xlab("Isolates") +
  ylab("Total Reads") +
  ggtitle(paste0("Total Reads of ",samples_or_project_name," samples from ",seqCompany," (",seqData_trimmed_or_not,")"))

ggsave("TotalReads.png",dpi = 600, width = 8, height = 6, units = "in")

###----> Mean Read length

ggplot(stats[order(pmax(stats$MeanReadLength_Pre_Trimming, stats$MeanReadLength_Post_Trimming)),], aes (x=factor(Isolate, levels=Isolate), MeanReadLength)) +  
  geom_point(aes (y= MeanReadLength_Pre_Trimming, color = "MeanReadLength_Pre_Trimming"), size=1) + 
  geom_point(aes(y=MeanReadLength_Post_Trimming, color = "MeanReadLength_Post_Trimming"), size=1) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size=5),legend.position = "top") + #Rotating and spacing axis labels
  expand_limits(x = c(0, NA), y = c(0, NA)) +
  xlab("Isolates") +
  ylab("Mean Read Length ") +
  ggtitle(paste0("Mean Read Length of ",samples_or_project_name," samples from ",seqCompany," (",seqData_trimmed_or_not,")"))

ggsave("MeanReadLength.png",dpi = 600, width = 8, height = 6, units = "in")

  
###----> Total Bases

ggplot(stats[order(pmax(stats$TotalBases_Pre_Trimming, stats$TotalBases_Post_Trimming)),], aes (x=factor(Isolate, levels=Isolate), TotalBases)) +  
  geom_point(aes (y= TotalBases_Pre_Trimming, color = "TotalBases_Pre_Trimming"), size=1) + 
  geom_point(aes(y=TotalBases_Post_Trimming, color = "TotalBases_Post_Trimming"), size=1) +
  theme_bw() +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size=5),legend.position = "top") + #Rotating and spacing axis labels
  expand_limits(x = c(0, NA), y = c(0, NA)) +
  scale_y_continuous(labels = unit_format(unit = "MB", scale = 1e-6)) + ## need to load "scales" library - very readable
  xlab("Isolates") +
  ylab("Total Bases") +
  ggtitle(paste0("Total Bases of ",samples_or_project_name," samples from ",seqCompany," (",seqData_trimmed_or_not,")"))

ggsave("TotalBases.png",dpi = 600, width = 8, height = 6, units = "in")

```


### Step 4: SPADes assembly 

The resulting cleaned data from illumina is passed as input to `SPADes`. 

```{r assembly, engine = 'bash', eval = FALSE}
cd /storage/data/DATA4/analysis/27_Paeruginosa_183_samples_Shawn_with_16_Envsamples/2_CleanData

time for d in $(ls -d */ | grep -v 'multiqc'); do echo $d; subdir=`echo -n $d | tr -d "/"`; cd $subdir; R1=`ls *_1_bbmap_adaptertrimmed.fq `; R2=`ls *_2_bbmap_adaptertrimmed.fq`; echo "$R1 $R2"; spades.py --pe1-1 "$R1" --pe1-2 "$R2" -o "$subdir"_spades --careful -t 48 >>log_spades; cd ..; done

mkdir ../3_SPAdes
mv */*_spades/ ../3_SPAdes/

#Copy and rename the assemblies
cd ../3_SPAdes/

for d in $(ls *_spades/contigs.fasta); do echo $d; prefix=`echo "$d" | cut -f1 -d "/"`; cp "$d" "$prefix"_contigs.fasta; done

mkdir ../4_SPAdes_Assemblies
mv *contigs.fasta ../4_SPAdes_Assemblies/
cd ../4_SPAdes_Assemblies/

Filter for contigs >1kb
------------------------
used bioawk installed in server

$ for d in $(ls *.fasta | sed 's/_contigs.fasta//g' ); do echo $d; /storage/apps/bioawk/bioawk -c fastx 'length($seq) >=1000 {print "\>"$name"\n"$seq}' "$d"_contigs.fasta >$d.gte1kb.contigs.fasta; done

mkdir gteq1kb
mv *gte1kb* gteq1kb/
mkdir full_assembly
mv *.fasta full_assembly/
```

### Step 5: Assembly Downstream - MLST

```{r mlst, engine = 'bash', eval = FALSE}

cd gteq1kb

# This is plotting the stats
statswrapper.sh *.fasta >gteq1kb_assemblies_summary_stats.txt

mlst *contigs.fasta | grep ".fasta" >>log_mlst_2.19

mkdir ../../5_MLST
mv log_mlst* ../../5_MLST/

```

### Step 6: Assembly Downstream - CGE - Resfinder4.1

Although Resfinder is part of CGE-BAPS, only version 2.1 is used for finding the resistance genes. So, we downloaded the resfinder4.1 version locally on the server which is fast to detect resistance genes.

Running resfinder on all assembly fasta - default setting is coverage length of the sequence is 60% and identity 80%

```{r resfinder4.1, engine = 'bash', eval = FALSE}

time for d in $(ls *.fasta); 
do 
echo $d; 
time python3 /storage/apps/resfinder4.1/run_resfinder.py -o "$d"_resfinder_outdir -s "Other" -l 0.6 -t 0.8 --acquired -ifa $d & 
done

```

### Step 7: Generate cge_resistance similar to BAPs format

Here ResFinder_results_tab.txt results are taken from each directory | remove spaces | unwanted columns removed | trimmed unnecessary texts | filtered the AMR genes which has coverage and identity >=80% | extracted unique records | collapsed the genes for each of the sample


```{r resfinder4.1_dataparse, engine = 'bash', eval = FALSE}

$ awk '{print FILENAME "\t" $0}' */ResFinder_results_tab.txt  | tr ' ' '_' | cut -f1-9 -d "       " | sed -e 's/_spades.gte1kb.contigs.fasta_resfinder_outdir\/ResFinder_results_tab.txt//g' | awk '$3>=80 && $5>=80'| awk '{print $1 "\t" $2}' | sort -u | grep -v 'Resistance_gene' | groupBy -g 1 -c 2 -o collapse  >cge_resfinder_resistome.tab

# head(cge_resfinder_resistome.tab) 

#A1089	blaACT-16,blaOXA-181,fosA,mdf(A),OqxA,OqxB,qnrS1
#A1416	aac(6')-Ib-cr,aadA16,ARR-3,blaCTX-M-15,blaSHV-11,...

mkdir ../../6_CGE
mv *resfinder* ../../6_CGE/
```

### Step 7a: Plot resistome

Here ResFinder_results_tab.txt results are taken from each directory | remove spaces | unwanted columns removed | trimmed unnecessary texts | filtered the AMR genes which has coverage and identity >=80% | extracted unique records | collapsed the genes for each of the sample


```{r plot_resistome, fig.width=12, fig.height=10, message=FALSE}


```



### Step 8: Core genome analysis - Prokka/Roary


```{r coregenome, engine = 'bash', eval = FALSE}
cd Serratia_marcescens

# created a list file with serratia unicycler assemblies 
wc -l Serratia_marcescens_unicycler_assemblies.list

# Copy the softlinks of the Serratia assembly fasta files 
$ for d in $(cat Serratia_marcescens_unicycler_assemblies.list ); do echo $d; ln -s /storage/data/DATA4/analysis/30_Aqueous_Env_11F_data_analysis/4_Unicycler_Assemblies/gteq1kb/$d .; done

# Since we are not performing coregenome by ST, we do not need to segregate fasta files by ST

# Run Prokka
# A duplicate local version Prokka called "Prokka2" is created in same folder to circumvent maxcontig length ID error
# by increasing from 37 (prokka script) to 50 (prokka2 script)

$ time for d in $(ls *.fasta | sed 's/_spades.gte1kb.contigs.fasta//g'); 
do 
echo "$d"; 
prokka2 --force --cpus 32 --outdir "$d"_prokka_out --prefix "$d" "$d"_spades.gte1kb.contigs.fasta >>prokka_log 2>>prokka_error; 
done

mkdir ../../7_Prokka
mv *prokka* ../../7_Prokka
# Added previously run prokka output folders for 179 PAE samples into this folder with 16 environmental samples

cd ../../7_Prokka
mkdir ../8_Roary
cp */*.gff ../8_Roary/

#9. Roary
########

cd ../8_Roary/
time roary -e --mafft -g 60000 -p 48 -cd 95 *.gff

real	46m53.349s
user	318m27.833s
sys	13m4.663s



# --> Create a list of all the samples according to Species (uusually it is by ST type) to be SNP validated
# If according to STs, then this oneliner generates list of files with the sample IDs according to the ST assigned by MLST tool
# Make sure there is a tab delimted space when using the fgrep

#for d in $(cat ../5_MLST/log_mlst | awk '{print $3}' | sort -u);do
#num=`grep -P "\t$d\t" ../5_MLST/log_mlst | wc -l`;
#grep -P "\t$d\t" ../5_MLST/log_mlst | awk '{print $1 "\t" $3}' | sed 's/_spades.*//g' >list_PAE_ST"$d"_"$num"samples;
#done

#$ cp list_PAE_ST*samples ../11_SNP_Validation/
mkdir ../9_SNP_Validation/
cp core_gene_alignment.aln  ../9_SNP_Validation/
ls *.gff | sed 's/.gff//g' >../9_SNP_Validation/list_PAE_266samples

#10. SNP Validation
###################

cd ../9_SNP_Validation/

cp /storage/apps/SNP_Validation_Scripts/tools/snpvalidation.sh .

In the snpvalidation.sh script, change the path of the directory for R1 and R2 to run snippy. 

--R1 /storage/data/DATA4/analysis/27_Paeruginosa_183_samples_Shawn_with_16_Envsamples_31_Jeanette_40_jocelynsamples/2_CleanData/$d/"$d"*_1_bbmap_adaptertrimmed.fastq 
--R2 /storage/data/DATA4/analysis/27_Paeruginosa_183_samples_Shawn_with_16_Envsamples_31_Jeanette_40_jocelynsamples/2_CleanData/$d/"$d"*_2_bbmap_adaptertrimmed.fastq

$ time ./snpvalidation.sh list_PAE_266samples

real	411m24.406s
user	3530m55.569s
sys	256m3.535s


11. Gubbins (We no more do padding before gubbins) &&  Calculate SNP difference
#################################################################################

# Once the snpvalidation is done, change directory into the PAE_ST308_179samples, and run gubbins

$ cd /storage/data/DATA4/analysis/27_Paeruginosa_183_samples_Shawn_with_16_Envsamples_31_Jeanette_40_jocelynsamples/9_SNP_Validation/PAE_266samples

mkdir Gubbins
cp core_gene_alignment_withConsensus_convertXtoN_convertgaptoLittleN_editedByGATK_renamed.fasta Gubbins/
cd Gubbins/

# Convert the consensus alignment into one liner fasta format

$ perl /storage/apps/SNP_Validation_Scripts/tools/convertFastatoOneLine.pl core_gene_alignment_withConsensus_convertXtoN_convertgaptoLittleN_editedByGATK_renamed.fasta

# Assuming consensus is on top remove consensus sequence from the above fasta file
sed '1,2d' core_gene_alignment_withConsensus_convertXtoN_convertgaptoLittleN_editedByGATK_renamed.fasta_OneLine.fasta >>core_gene_alignment_woConsensus_editedByGATK_renamed.fasta

# Gubbins - Filter recombination
$ time run_gubbins.py --prefix postGubbins --filter_percentage 100 --threads 48 core_gene_alignment_woConsensus_editedByGATK_renamed.fasta --verbose 

$ cp /storage/apps/SNP_Validation_Scripts/tools/dnaDist.r .

$ Rscript dnaDist.r 



#(Repeat the above steps in Gubbins for all the Species (or ST groups))

# Since it is too tiring to run each and every step, I created a shell script which run the above commands for the rest of the species. So we can run all the steps indivisually or run the below script if there are multiple Species and ST to deal with 

$ time bash runProkka_Roary_Gubbins_bySpecies.sh


12. Clustering 
###############

# We want to cluster all the isolates which have a SNP difference  with certain threshold (for this case : 0 SNPs, Therefore my awk has $NF==0)

# The following one-liner takes the SNP difference file | takes pair1,pair2,SNP Diff | removes double quotes if any | extract pairs satisfyingSNP Diff threshold (Here 0 SNPs) | print rhe pair1, pair2 to a file

$ cat postGubbins.filtered_polymorphic_sites_melt_sorted.csv | awk -F',' '{print $2"\t"$3"\t"$4}' |  tr -d "\"" | awk '$NF==0' | awk '{print $1"\t"$2}' >pairsFile.list

# The Clustering Script will recruit the isolates if any of the isoaltes has link with another to the group

$ cp /storage/apps/PlasmidSeeker/Databases_aftr_04122019/ECCMID_2020/Deduplicated_Fasta_for_PlasmidSeeker_DB/ClustrPairs_v3.pl .

$ perl ClustrPairs_v3.pl pairsFile.list 

# Looking at thresholds

for d in {0..11}; do cat postGubbins.filtered_polymorphic_sites_melt_sorted.csv | awk -F',' '{print $2"\t"$3"\t"$4}' |  tr -d "\"" | awk -v cutoff="$d" '$NF<=cutoff' | awk '{print $1"\t"$2}' >pairsFile__snpscutoff__"$d".list; clusters=`perl /storage/apps/SNP_Validation_Scripts/tools/ClustrPairs_v3.pl pairsFile__snpscutoff__"$d".list | fgrep -c "Cluster:"`; perl /storage/apps/SNP_Validation_Scripts/tools/ClustrPairs_v3.pl pairsFile__snpscutoff__"$d".list >Clusters__snpscutoff__"$d".list; samples=`grep -v 'Cluster:' Clusters__snpscutoff__"$d".list | wc -l`; echo "At SNP cutoff $d - Clusters: $clusters - with $samples samples";  done


13. Visualization of tree using Figtree: 
#########################################



$ java -jar /storage/apps/FigTree_v1.4.3/lib/figtree.jar

File -> Open -> postGubbins.final_tree.tre

```

### Step temp: PlasClass
```
$ time python train.py -c /storage/data/DATA4/analysis/27_Paeruginosa_183_samples_Shawn/16_Check_presence_of_ICETn43716385/NC_002516.fasta -p /storage/data/Storage/plsdb/plsdb.fasta -o /storage/data/DATA4/analysis/27_Paeruginosa_183_samples_Shawn/19_PlasClass_PAE -n 48

```


### Step 8b: Visualizing SNPs for various Species


```{r snp1, fig.width=12, fig.height=8, message=FALSE}
setwd("/data02/Analysis/Projects/8_Aqueos_samples/9_SNP_matrix/")

###This section draws cumulative SNP difference of the dataset from melt sorted file

# library(splitstackshape)
# library(dplyr)
# library(reshape2)
# library(ggplot2)
# library(gplots)
# 
# half_snp_table <- read.table("postGubbins.filtered_polymorphic_sites_SM_half_triangle.csv",sep = ",", header = TRUE)
# head(half_snp_table)
# colnames(half_snp_table) <- c("No","Sample1","Sample2","SNPDiff")
# head(half_snp_table)
# tail(half_snp_table)
# 
# half_snp_table <- na.omit(half_snp_table)
# tail(half_snp_table)
# #half_snp_table[1] <- NULL  #remove 1st column
# class(half_snp_table)
# head(half_snp_table)
# tail(half_snp_table)
# 
# half_snp_table$CSNPDiff <- cumsum(half_snp_table$SNPDiff)
# head(half_snp_table)
# tail(half_snp_table)
# 
# library(reshape)
# df_snpdifflteq10 <- half_snp_table %>% filter(SNPDiff <=15)
# head(df_snpdifflteq10)
# 
# ggplot(df_snpdifflteq10) + geom_point(aes(x = No, y = CSNPDiff), color = 'blue', size = 0.1) +
# scale_x_continuous(breaks = scales::pretty_breaks(n = 30), limits = c(0, NA)) +
# scale_y_continuous(breaks = scales::pretty_breaks(n = 20), limits = c(0, NA)) +
# ggtitle("P.aeruginosa ST308 SNPDiff Cumulative Sum (SNP Cutoff <=1)") + # for the main title +
# xlab("Pair") + # for the x axis label
# ylab("Cumulative SNP Diff Sum")
# ggplot(half_snp_table) + geom_point(aes(x = No, y = CSNPDiff), color = 'blue', size = 0.1) + scale_x_continuous(breaks = scales::pretty_breaks(n = 30), limits = c(0, NA)) + scale_y_continuous(breaks = scales::pretty_breaks(n = 20), limits = c(0, NA))

####################END

#temp = list.files(pattern="*_full_triangle.csv")
#temp

SNPDiff_MatPlot <- function(fileName, imageName, cellwidth_tmp, cellheight_tmp, Species_ST) 
{

snp_table <- read.table(fileName,sep = ",", header = TRUE)

#snp_table <- read.table(temp[i],sep = ",", header = TRUE)
head(snp_table)
colnames(snp_table) <- c("No","Sample1","Sample2","SNPDiff")
head(snp_table)
tail(snp_table)

complete_snp_table <- na.omit(snp_table)
tail(complete_snp_table)
complete_snp_table[1] <- NULL 
class(complete_snp_table)
complete_snp_table

# Unsorted SNP Heatmap
# ggplot(complete_snp_table, aes(Sample1,Sample2, fill = SNPDiff, alpha = SNPDiff)) +
#   geom_tile(colour = "black") +
#   geom_text(aes(label=SNPDiff)) +
#   scale_alpha_identity(guide = "none") +
#   coord_equal(expand = 0) +
#   theme_bw() +
#   theme(panel.grid.major = element_blank(),
#         axis.text.x = element_text(angle = 90, hjust = 1)) +
#   ggtitle("Wenjian data ST11 SNPDiff matrix") + # for the main title
#   xlab("Samples") + # for the x axis label
#   ylab("Samples")+ scale_fill_gradient(low = "white", high = "tomato3")

## To generate a heatmap with SNP Difference values and sorted by the SNP values we will use pheatmap

library(pheatmap)

wide_complete_snp_table <- acast(complete_snp_table, Sample1~Sample2, value.var="SNPDiff")
head(complete_snp_table)
head(wide_complete_snp_table)

#colfunc <- colorRampPalette(c("white","tomato3"))
colfunc <- colorRampPalette(c("#ffd662ff","#24868EFF","tomato3"))


# pheatmap(wide_complete_snp_table, display_numbers = T,number_format = "%s",color=colfunc(10), fontsize = 6, main = "P.aeruginosa ST308 SNPDiff matrix")
# 
# pheatmap(wide_complete_snp_table, display_numbers = T,number_format = "%s",color=colfunc(10),cellwidth = 15, cellheight = 12, fontsize = 6, main = "P.aeruginosa ST308 SNPDiff matrix") 
# 
# pheatmap(wide_complete_snp_table, display_numbers = T,number_format = "%s",color=colfunc(10),cellwidth = 15, cellheight = 12, fontsize = 6, border_color = "black", main = "P.aeruginosa ST308 SNPDiff matrix")

save_pheatmap_png <- function(x, filename, width=2000, height=2000, res = 200) {
  png(filename, width = width, height = height, res = res)
  grid::grid.newpage()
  grid::grid.draw(x$gtable)
  dev.off()
}

#pheatmap(wide_complete_snp_table, display_numbers = T,number_format = "%s",color=colfunc(30),cellwidth = cellwidth_tmp, cellheight = cellheight_tmp, fontsize = 8, border_color = "gray30", number_color = "black", main = paste0(Species_ST, " SNPDiff matrix"))

SNPDiff_hm <- pheatmap(wide_complete_snp_table, display_numbers = T,number_format = "%s",color=colfunc(30),cellwidth = cellwidth_tmp, cellheight = cellheight_tmp, fontsize = 8, border_color = "gray30", number_color = "black", main = paste0(Species_ST, " SNPDiff matrix"))

SNPDiff_hm
#save_pheatmap_png(SNPDiff_hm, "EntSp_SNPDiff_Matrix_pheatmap.png")
save_pheatmap_png(SNPDiff_hm, paste0(imageName,"_SNPDiff_Matrix_pheatmap.png"))
}

SNPDiff_MatPlot("postGubbins.filtered_polymorphic_sites_SM_full_triangle.csv", "SM", cellwidth_tmp=15, cellheight_tmp=12, Species_ST="Serratia marcescens") 

SNPDiff_MatPlot("postGubbins.filtered_polymorphic_sites_KP_full_triangle.csv", "KP", cellwidth_tmp=25, cellheight_tmp=25, Species_ST="Klebsiella pneumoniae") 

SNPDiff_MatPlot("postGubbins.filtered_polymorphic_sites_EC_full_triangle.csv", "EC", cellwidth_tmp=25, cellheight_tmp=25, Species_ST="Escherichia coli") 

SNPDiff_MatPlot("postGubbins.filtered_polymorphic_sites_EntSp_full_triangle.csv", "EntSp", cellwidth_tmp=25, cellheight_tmp=25, Species_ST="Enterobacter Species") 

```

### Step 8c: Gubbins Tree

Lets plot gubbins tree. Can simply plot the tree using Figtree

```{r gubbins_tree1, fig.width=12, fig.height=8, message=FALSE}

# $ perl newNames_newick.pl newNames.txt postGubbins.SM.final_tree.tre >postGubbins.SM.renamed.final_tree.tre

library(ape)
library(RCurl)

setwd("/data02/Analysis/Projects/8_Aqueos_samples/11_GubbinsTree")

list.files()
# a tree was made using code in a previous blog post. 
# see here: http://rforbiochemists.blogspot.co.uk/2017/01/visualizing-kinome-in-r-simple-tree.html
# it's here on github
#link <- "https://raw.githubusercontent.com/brennanpincardiff/RforBiochemists/master/phyloTrees/kinaseTree_20161221"

# this will download it into R. 
# using read.tree() from ape package
PlotApePhyloTree <- function(newickFile, organism, width=30, height=30){
dev.new()
tree <- read.tree(file = newickFile)

# this tree looks quite nice in my opinion and is the starting point of this blog 
# plot(tree, "u", 
#      use.edge.length = FALSE,
#      show.tip.label = FALSE,
#      edge.color = "red")
# but lacking colours for the groups and any labels...





# it seems more traditional to draw trees from left to right. 
# plot(tree, 
#      use.edge.length = FALSE,
#      show.tip.label = FALSE,
#      edge.color = "red")


# with some names... this is slow to draw due to the names 
# plot(tree, 
#      use.edge.length = FALSE,
#      edge.color = "red",
#      cex = 0.25)



# and you can't read the names because there are >500




# to customise this tree in a way we want we need to understand a little more about trees
# we can find out more about an object by writing the name
tree
# "Phylogenetic tree with 516 tips and 514 internal nodes"

# by using the class() function
class(tree)
# "phylo"

# or by using the str() structure function
str(tree) 
# "List of 4"
# this list includes $edge, $Nnode, $ tip.label and $edge.length
# the tree$tip.label includes family designation
tree$tip.label  # 516 of these

# from the Science paper, we have seven kinase families:
# kinase categories... TK, TKL, STE, CK1, AGC, CAMK, CMGC
# with the following colours
# "red", "green", "paleblue", "orange", "yellow", "purple", "pink", "green" 


# by using the grep()function on the tree$tip.label part of the object
# we can find the tip labels that include "TK/" - i.e. tyrosine kinases
grep("_C1", tree$tip.label)  # gives a list of numbers with "TK/" in tip label
length(grep("_C1", tree$tip.label))
# thus there are 94 tip labels with that are designated TK (not TKL tyrosine kinase like)

# make a vector for each tip.label called tipcol with black on all of these...
tipcol <- rep('black', length(tree$tip.label))
tipcol
# make a vector with our list of kinase categories
kinaseCats <- c("_C1", "_C2", "_C3", "_C4", "_C5", "_C6","_C7")

# make a vector of color we want:
colorsList <-c("red", "darkgreen", "blue", "orange", "navyblue", "purple", "maroon", "green")

# replace colours where grep gives "TK" as red, etc in a loop
for(i in 1:length(kinaseCats)){
  tipcol[grep(kinaseCats[i], tree$tip.label)] <- colorsList[i]
}

# plot with edge length false to see nodes better
#plot(tree)

# plot(tree, 
#      tip.color=tipcol, 
#      cex = 0.75)

#plot.phylo(tree,tip.color=tipcol,cex = 0.75,edge.width = 2)

#plot.phylo(tree,tip.color=tipcol,cex = 0.75, main = "Serratia marcescens Gubbins Tree", font=2)

plot.phylo(tree,
           tip.color=tipcol,
           cex = 0.75,
           use.edge.length = TRUE,
           main = paste0(organism," Gubbins Phylogenetic Tree"), 
           font=2)

library(extrafont)
#font_import() # import fonts - takes some time
loadfonts()
#png("phylo1.png",width = 800, height = 800)
pdf(paste0(organism,"_Gubbins_phylotree.pdf"), family="Times New Roman", width=width, height=height)

plot.phylo(tree,
           tip.color=tipcol,
           cex = 0.3,
           use.edge.length = TRUE,
           main = paste0(organism," Gubbins Phylogenetic Tree"), 
           font=2)
dev.off()

#graphics.off()

}

PlotApePhyloTree("postGubbins.EC.renamed.final_tree.tre", "Escherichia coli", width=5, height=5)

PlotApePhyloTree("postGubbins.EntSp.renamed.final_tree.tre", "Enterobacter Species", width=5, height=5)
PlotApePhyloTree("postGubbins.KP.renamed.final_tree.tre", "Klebsiella pneuomoniae", width=5, height=5)
PlotApePhyloTree("postGubbins.SM.renamed.final_tree.tre", "Serratia marcescens", width=5, height=5)



```



### Step 9: plasmid data analysis

Kraken2 for species calling on adapter trimmed reads of illumina. 


```{r plasmid_replicon1, fig.width=12, fig.height=8, message=FALSE}

#Step 1: Run mobtyper
#I wrote a shell script with mobtyper which breaks an assembly into single contig fasta file and run mobtyper on each contig seperately. By doing so, we can know which specific contig has what Inc group.

##Bash
#time ./mob_typer_OnEachContig.sh &
#$ fgrep '' *_unicycler.gte1kb.contigs.fasta_Mob/*_unicycler.gte1kb.contigs.fasta.Mob_report.txt | awk -F/ '{print $2}' | tr ':' '       ' | awk '{print $1 "\t" $2 "\t" $6 "\t" $8 "\t" $14}' | sed 's/_unicycler.gte1kb.contigs.fasta.Mob_report.txt//g' | grep -v 'file_id' >Mob_Summary.txt


# Extracted below columns from the Mob typer output for all the samples
# SampleName	file_id	rep_type(s)	relaxase_type(s)	PredictedMobility

### Step 3: Generate R plots
#From the assembly.fasta, the headers were extracted and an text file **Cmp_RawFilt_Assemblies.txt** is created which is used for plotting using R ggplot function.
```

##### Let's Mob

```{r plasmid_replicon2, fig.width=12, fig.height=10, message=FALSE}
setwd("/data02/Analysis/Projects/8_Aqueos_samples/8_Mob")

library(ggplot2)
library(dplyr)
library(RColorBrewer)
library(ggthemes)
library(paletteer)
library(stringr)
library(wesanderson)
library(scales)
library(kableExtra)


MobDF <- read.table("Mob_Summary.txt",sep = "\t", header = FALSE)
head(MobDF) %>%
  kbl(caption = "Head sample output") %>%
   kable_classic_2(full_width = F)

colnames(MobDF) <- c("sample_name", "contig",	"replicon_type",	"relaxase_type", "PredictedMobility")
head(MobDF) %>%
  kbl(caption = "Head sample output") %>%
   kable_classic_2(full_width = F)


MobDF_mod <- MobDF %>%
  mutate(interim = str_extract(contig, "_length=.*_depth")) %>%
  mutate(length = str_extract(interim, "[0-9]+")) %>%
  select(-interim)

head(MobDF_mod) %>%
  kbl(caption = "Head sample output") %>%
   kable_classic_2(full_width = F)

tail(MobDF_mod) %>%
  kbl(caption = "Head sample output") %>%
   kable_classic_2(full_width = F)
```

ST information is not available in Mobtyper results. SO adding that information from CGE results for each the sample.

```{r plasmid_replicon3, fig.width=16, fig.height=10, message=FALSE}
library(kableExtra)
# lets add ST from an external file using vlookup

setwd("/data02/Analysis/Projects/8_Aqueos_samples/8_Mob")

ST <- read.table("cge_resistome_species_ST.txt",sep = "\t", header = TRUE, quote="")
head(ST) %>%
  kbl(caption = "Head sample output") %>%
   kable_classic_2(full_width = F)

# Vlookup in R using the dplyr package
MobDF_mod_ST <- as.data.frame(MobDF_mod %>%
                                right_join(ST, by = 'sample_name') %>% 
                                select(-resistance) %>% 
                                mutate(Species_ST = paste0(Species, "_ST", ST)
                                       )
                              )
head(MobDF_mod_ST) %>%
  kbl(caption = "Head sample output") %>%
   kable_classic_2(full_width = F)

# how many samples have plasmids?
sample_replicon_count <- as.data.frame(MobDF_mod_ST %>%
  filter(!replicon_type == "-")  %>%
 # head() %>%
  select(sample_name,replicon_type,ST,Species_ST) %>%
  unique() %>%
  arrange(replicon_type) %>%
  group_by(sample_name) %>%
  mutate(replicon = paste0(replicon_type, collapse = ",")) %>%
  select(-replicon_type) %>%
  mutate(count = n()) %>%
  unique() %>%
  arrange(count,replicon))

head(sample_replicon_count) %>%
  kbl(caption = "Head sample output") %>%
   kable_classic_2(full_width = F)

write.csv(sample_replicon_count,"sample_repliconCount.csv")
```

Let us plot how many of Replicons each sample carries from the dataset.

```{r plasmid_replicon4, fig.width=16, fig.height=10, message=FALSE}
sample_replicon_count$replicon <- as.factor(sample_replicon_count$replicon)
sample_replicon_count$sample_name <- as.factor(sample_replicon_count$sample_name)

c25 <- c(
  "dodgerblue2", "#E31A1C", # red
  "green4",
  "#6A3D9A", # purple
  "#FF7F00", # orange
  "black", "gold1",
  "skyblue2", "#FB9A99", # lt pink
  "palegreen2",
  "#CAB2D6", # lt purple
  "#FDBF6F", # lt orange
  "maroon",
  "gray70", "khaki2",
  "orchid1", "deeppink1", "blue1", "steelblue4",
  "darkturquoise", "green1", "yellow4", "yellow3",
  "darkorange4", "brown"
)


# Replicon Count of Samples by Sequence Type
ggplot(sample_replicon_count,aes(reorder(x=sample_name,count),y=count,color=replicon))+
  geom_point(size = 1) +
  facet_grid(~Species_ST, scales = "free", space = "free") +
  #scale_color_brewer(palette="Set3") +
  scale_color_manual(values=c25)+
  #scale_color_manual(values=wes_palette(n=12,  name = "Darjeeling1", type = "continuous"))+
  #scale_shape_manual(values = c(0,1,2,3,4,5,6,7,8,9,10,11,12)) +
  theme(panel.grid.major = element_blank(),
        axis.text.x = element_text(angle = 90, hjust = 1, size = 6),
        axis.text.y = element_text(angle = 0, hjust = 1, size = 7),
        legend.position = "top", legend.text=element_text(size=5),strip.text.y.right = element_text(angle = 0), strip.text.x.top = element_text(angle = 90)) +
  labs(x="Sample Name",
       y="Replicon Count",
       title = "Replicon Count of Samples by Sequence Type")

ggsave("MobTyper_Replicon_count_perIsolate.png",dpi = 300, width = 20, height = 10, units = "in")

```


```{r plasmid_replicon5, fig.width=16, fig.height=18, message=FALSE}

library(kableExtra)
# Replicon Count of Samples by Sequence Type
head(sample_replicon_count) %>%
  kbl(caption = "Head sample output") %>%
   kable_classic_2(full_width = F)

MobDF_mod_ST_table <- as.data.frame(sample_replicon_count %>%  group_by(Species_ST) %>% count(replicon))
head(MobDF_mod_ST_table) %>%
  kbl(caption = "Head sample output") %>%
   kable_classic_2(full_width = F)

ggplot(MobDF_mod_ST_table,aes(x=replicon,y=n,fill=replicon))+
  geom_bar(stat = "identity", width=0.4) +
  facet_grid(~Species_ST, scales = "free", space = "free") + 
  #scale_color_brewer(palette="Set3") +
  scale_color_manual(values=wes_palette(n=12,  name = "Darjeeling1", type = "continuous"))+
  #scale_shape_manual(values = c(0,1,2,3,4,5,6,7,8,9,10,11,12)) +
  geom_text(aes(label = n), vjust = -0.09, hjust = 0.07, position = position_dodge(width=0.9),  size=3)+
  theme(panel.grid.major = element_blank(),
        axis.text.x = element_text(angle = 45, hjust = 1, size = 9),
        axis.text.y = element_text(angle = 0, hjust = 1, size = 9),
        legend.position = "none", legend.text=element_text(size=7),
        strip.text.y.right = element_text(angle = 0), strip.text.x.top = element_text(angle = 90),
        plot.margin = unit(c(1,1,1,9), "cm")) +
  labs(x="Replicon combinations",
       y="Replicon Count",
       title = "Replicon Count of Samples by Sequence Type") #+ 
   # + coord_flip()

ggsave("MobTyper_Replicon_count_perIsolate2.png",dpi = 300, width = 15, height = 11.5, units = "in")

```

From the above we can understand all of our Aqueous samples have a XXX replicon and XXX replicon.

### Step 10: Gene Arrow Maps on the CP genes

Kraken2 for species calling on adapter trimmed reads of illumina. 

```{r prokka_gff_parse, engine = 'bash', eval = FALSE}

# SM

$ fgrep "" */*.gff | fgrep 'CDS' | sed 's/;product/\'$'\tproduct/' | sed 's/ /_/g' | awk '{print $1,$4,$5,$3,$7,$NF}' | fgrep -v 'hypothetical_protein' | sed 's/_prokka.*.gff:/\'$'\t/g' | sed 's/ + / 1 /g' | sed 's/ - / -1 /g' | sed 's/ /\'$'\t/g' >SM_Prokka_coordinates_files.gff

```

```{r gene_arrow_maps, fig.width=16, fig.height=18, message=FALSE}

setwd("/data02/Analysis/Projects/8_Aqueos_samples/10_GeneArrowMaps")
#getwd()
#files()

#Let us load the coordinates_files.gff into a dataframe

coords <- read.table("SM_Prokka_coordinates_files.gff",header = FALSE, sep = "\t", quote = "")
head(coords)

coords <- coords %>% 
unite(molecule, V1, V2, sep = "_", remove = TRUE)

names(coords) <- c("molecule","start","end","source","direction","gene")
head(coords)

coords2 <- within(coords, rm(source))
coords2$strand[coords$direction=="1"] <- "forward"
coords2$strand[coords$direction=="-1"] <- "reverse"
head(coords2)

write.csv(coords2,file="SM_GeneArrowMaps.csv")

#load this in gamR or run ggmaps
```

### Step 11: Plotting CP genes plasmids by sizes

Wrote a script `extract_putativeCircPlasmids.sh` to extract all putative plasmids from the assemblies

```{r plasmid_exploration_bash, engine = 'bash', eval = FALSE}
bash extract_putativeCircPlasmids.sh

# Segregate Plasmids by Species
mkdir Serratia_Plasmids

for d in $(cat ../../8_Prokka/Serratia_marcescens/Serratia_marcescens_unicycler_assemblies.list | sed 's/_unicycler.gte1kb.contigs.fasta//g'); do echo $d; mv "$d"_* Serratia_Plasmids/

for d in $(cat ../../8_Prokka/Klebsiella_pneumoniae/Klebsiella_pneumoniae_unicycler_assemblies.list | sed 's/_unicycler.gte1kb.contigs.fasta//g'); do echo $d; mv "$d"_* Klebsiella_Plasmids/; done

for d in $(cat ../../8_Prokka/Enterobacter_Species/Enterobacter_Species_unicycler_assemblies.list | sed 's/_unicycler.gte1kb.contigs.fasta//g'); do echo $d; mv "$d"_* EntSp_Plasmids/; done

mv *.fa Escherichia_Plasmids/

ls -1 */*.fa | sed 's/_Plasmids\//,/g' >Plasmid_length_distr_by_species.csv
```

Plotting CP plasmids by length and species

```{r plasmid_exploration_plot1, fig.width=16, fig.height=18, message=FALSE}

setwd("/data02/Analysis/Projects/8_Aqueos_samples/12_Plasmids_length_distr_by_species")

#Let us load the coordinates_files.gff into a dataframe
library(dplyr)
library(stringi)
library(ggplot2)
library(ggridges)

df <- read.csv("Plasmid_length_distr_by_species.csv",header = FALSE, sep = ",")
head(df)

df <- df %>% 
  mutate(interim = str_extract(V2, "_len_.*_circ")) %>% 
  mutate(length = str_extract(interim, "[0-9]+")) %>% 
  select(-interim) %>% 
  mutate(sample_name = str_extract(V2, "[^_]+")) #%>% head()
  
head(df)
names(df) <- c("species","contig", "CPGene", "length","sample_name")
head(df)
tail(df)

df$length <- as.numeric(df$length)

# Do not need histogram for now
# ggplot(df, aes(x=length,fill=species))+
#   geom_histogram(stat="bin", position="dodge") +
#   scale_x_continuous(labels = scales::comma,breaks = scales::pretty_breaks(n=40))+
#   facet_grid(species~.) + 
#   stat_bin(aes(y=..count.., label=ifelse(..count..==0,"",..count..)), geom="text", vjust=-0.2, size = 3) +
#   theme(panel.grid.major = element_blank(),
#         axis.text.x = element_text(angle = 90, hjust = 1, size = 9),
#         axis.text.y = element_text(angle = 0, hjust = 1, size = 9),
#         legend.position = "none", legend.text=element_text(size=7),
#        # strip.text.y.right = element_text(angle = 0), strip.text.x.top = element_text(angle = 90),
#        # plot.margin = unit(c(1,1,1,9), "cm")
#        ) +
#   labs(x="Replicon combinations",
#        y="Replicon Count",
#        title = "Replicon Count of Samples by Sequence Type") 

head(df)

#Plot X axis scale as continuous 
 
# ggplot(df, aes(length, sample_name)) +
#   geom_point(aes(color = CPGene),size=1) +
#   facet_grid(species~.,scales = "free_y") +
#   scale_x_continuous(labels = scales::comma, breaks = scales::pretty_breaks(n=30)) +
#   scale_color_manual(values = c("KPC-2" = "darkblue", "OXA-181" = "#FC4E07", "OXA-48" = "#00AFBB", "No_CPGene" = "#000000")) +
#   theme(panel.grid.major = element_blank(),
#         axis.text.x = element_text(angle = 90, hjust = 1, size = 9),
#         axis.text.y = element_text(angle = 0, hjust = 1, size = 7),
#         legend.position = "right", legend.text=element_text(size=7),
#        # strip.text.y.right = element_text(angle = 0), strip.text.x.top = element_text(angle = 90),
#         plot.margin = unit(c(1,1,1,1), "cm")
#        ) +
#   labs(x="Plasmid size",
#        y="Count of Plasmids",
#        title = "Plasmid Count of Aqueous Samples by Species") 

#Plot X axis scale as discrete

ggplot(df, aes(factor(length), sample_name)) +
  geom_point(aes(color = CPGene),size=1.5) +
  facet_grid(species~.,scales = "free_y") +
  #scale_x_continuous(labels = scales::comma, breaks = scales::pretty_breaks(n=30)) +
 # scale_color_manual(values = c("KPC-2" = "yellow", "OXA-181" = "purple", "OXA-48" = "#FC4E07", "No_CPGene" = "cyan4")) +
   scale_color_manual(values = c("KPC-2" = "purple", "OXA-181" = "#24868EFF", "OXA-48" = "#FC4E07", "No_CPGene" = "#ffd662ff")) +
  theme_bw() +
  theme(panel.grid.major = element_blank(),
        axis.text.x = element_text(angle = 90, hjust = 1, size = 9),
        axis.text.y = element_text(angle = 0, hjust = 1, size = 4),
        legend.position = "right", legend.text=element_text(size=7),
       # strip.text.y.right = element_text(angle = 0), strip.text.x.top = element_text(angle = 90),
        plot.margin = unit(c(1,1,1,1), "cm")
       ) +
  labs(x="Plasmid size",
       y="Sample",
       title = "CP & Non-CP plasmid presence in Aqueous Samples by Species") 
ggsave("CP_Plasmidoverview.png",dpi = 300, width = 16, height = 8, units = "in")

```


Plotting CP plasmids by length and species and patient

```{r plasmid_exploration_plot2, fig.width=16, fig.height=10, message=FALSE}

setwd("/data02/Analysis/Projects/8_Aqueos_samples/12_Plasmids_length_distr_by_species/Re_plotting_with_patients")

#Let us load the coordinates_files.gff into a dataframe
library(dplyr)
library(stringi)
library(ggplot2)
library(ggridges)

df <- read.csv("Plasmid_length_distr_by_species_and_patients.csv",header = TRUE, sep = ",")
head(df)

df <- df %>% 
  mutate(interim = str_extract(Contig, "_len_.*_circ")) %>% 
  mutate(Length = str_extract(interim, "[0-9]+")) %>% 
  select(-interim) 
  #%>% head()
  
head(df)
#names(df) <- c("species","contig", "CPGene", "length","sample_name")
head(df)
tail(df)

df$Length <- as.numeric(df$Length)

#head(df)
#df
#Plot X axis scale as continuous 
 
# ggplot(df, aes(length, sample_name)) +
#   geom_point(aes(color = CPGene),size=1) +
#   facet_grid(species~.,scales = "free_y") +
#   scale_x_continuous(labels = scales::comma, breaks = scales::pretty_breaks(n=30)) +
#   scale_color_manual(values = c("KPC-2" = "darkblue", "OXA-181" = "#FC4E07", "OXA-48" = "#00AFBB", "No_CPGene" = "#000000")) +
#   theme(panel.grid.major = element_blank(),
#         axis.text.x = element_text(angle = 90, hjust = 1, size = 9),
#         axis.text.y = element_text(angle = 0, hjust = 1, size = 7),
#         legend.position = "right", legend.text=element_text(size=7),
#        # strip.text.y.right = element_text(angle = 0), strip.text.x.top = element_text(angle = 90),
#         plot.margin = unit(c(1,1,1,1), "cm")
#        ) +
#   labs(x="Plasmid size",
#        y="Count of Plasmids",
#        title = "Plasmid Count of Aqueous Samples by Species") 

#Plot X axis scale as discrete

ggplot(df, aes(factor(Length), SampleName, color = factor(CPGene), shape = Source)) +
  geom_point(size = 2.5) +
  facet_grid(Species~.,scales = "free_y") +
  scale_color_manual(values = c("KPC-2" = "purple", "OXA-181" = "#24868EFF", "OXA-48" = "#FC4E07", "No_CPGene" = "#ffd662ff")) +
  scale_shape_manual(values = c(0, 4)) +
  theme_bw() + 
  theme(panel.grid.major = element_blank(),
                  axis.text.x = element_text(angle = 90, hjust = 1, size = 9),
                  axis.text.y = element_text(angle = 0, hjust = 1, size = 4),
                  legend.position = "right", legend.text=element_text(size=7),
                  # strip.text.y.right = element_text(angle = 0), strip.text.x.top = element_text(angle = 90),
                  plot.margin = unit(c(1,1,1,1), "cm")
  ) +
  labs(x="Plasmid size",
       y="Sample",
       title = "CP & Non-CP plasmid presence in Aqueous Samples by Species") 

ggsave("CP_Plasmidoverview.png",dpi = 300, width = 16, height = 8, units = "in")
 
df_cp <- df %>% filter(CPGene!="No_CPGene")

 ggplot(df_cp, aes(factor(Date), SampleName, color = factor(CPGene), shape = Source)) +
  geom_point(size = 3) +
  facet_grid(Species~.,scales = "free_y") +
  scale_color_manual(values = c("KPC-2" = "purple", "OXA-181" = "#24868EFF", "OXA-48" = "#FC4E07", "No_CPGene" = "#ffd662ff")) +
  scale_shape_manual(values = c(0, 4)) +
  theme_bw() + 
  theme(panel.grid.major = element_blank(),
        axis.text.x = element_text(angle = 90, hjust = 1, size = 9),
        axis.text.y = element_text(angle = 0, hjust = 1, size = 4),
        legend.position = "right", legend.text=element_text(size=7),
        # strip.text.y.right = element_text(angle = 0), strip.text.x.top = element_text(angle = 90),
        plot.margin = unit(c(1,1,1,1), "cm")
  ) +
  labs(x="Date of culture",
       y="Sample",
       title = "CP & Non-CP plasmid presence in Aqueous Samples by Species and Date") 

ggsave("CP_Plasmidoverview_byDate.png",dpi = 300, width = 16, height = 8, units = "in")




ggplot(df, aes(factor(Length), SampleName)) +
  geom_point(aes(color = CPGene, shape = Source),size=1) +
  facet_grid(Species~.,scales = "free_y") +
  #scale_x_continuous(labels = scales::comma, breaks = scales::pretty_breaks(n=30)) +
 # scale_color_manual(values = c("KPC-2" = "yellow", "OXA-181" = "purple", "OXA-48" = "#FC4E07", "No_CPGene" = "cyan4")) +
   scale_color_manual(values = c("KPC-2" = "purple", "OXA-181" = "#24868EFF", "OXA-48" = "#FC4E07", "No_CPGene" = "#ffd662ff")) +
   scale_shape_manual(limits = c("Patient", "Env"),      # manually select shape
                       values = c(f = 0, r = 2)) +
  theme_bw() +
  theme(panel.grid.major = element_blank(),
        axis.text.x = element_text(angle = 90, hjust = 1, size = 9),
        axis.text.y = element_text(angle = 0, hjust = 1, size = 4),
        legend.position = "right", legend.text=element_text(size=7),
       # strip.text.y.right = element_text(angle = 0), strip.text.x.top = element_text(angle = 90),
        plot.margin = unit(c(1,1,1,1), "cm")
       ) +
  labs(x="Plasmid size",
       y="Sample",
       title = "CP & Non-CP plasmid presence in Aqueous Samples by Species") 

ggsave("CP_Plasmidoverview.png",dpi = 300, width = 16, height = 8, units = "in")

```



### Step 12: Plasmid exploration

```{r plasmid_exploration2_bash, engine = 'bash', eval = FALSE}

#moved all serratia plasmids into a directory
for d in $(cat CP_Plasmids.list); do echo $d; mv ../"$d"_*.fa .; done
cat *.fa >Serratia_43_plasmids.fasta
dedupe.sh in=Serratia_43_plasmids.fasta out=Serratia_43_plasmids_dedup.fasta

# These plasmids exist in the non duplicates. Still looks very similar
# C492_4_len_67445_circ_plasmid
# C1347_5_len_67445_circ_plasmid
# C1363_5_len_68645_circ_plasmid
# C1379_4_len_67427_circ_plasmid
# C1401_4_len_68706_circ_plasmid
# C1288_4_len_68706_circ_plasmid
# C1292_6_len_67445_circ_plasmid
# C1239_4_len_68647_circ_plasmid
# C392_4_len_68222_circ_plasmid

# How different are they? Lets do some mash or kmer analysis

# makeblastdb -in Serratia_43_plasmids_dedup.fasta -dbtype nucl -out BlastDB/Serratia_Plasmids_DB -parse_seqids -logfile 

# blastn -db BlastDB/Serratia_Plasmids_DB -query Serratia_43_plasmids_dedup.fasta -outfmt '6 qseqid sseqid pident nident length mismatch gapopen qstart qend sstart send evalue bitscore qlen slen' -out Serratia_Plasmids_selfblastresults.txt

# Filtering with 99% cov and 99% identity
# $ awk '{print FILENAME"\t"$0}' Serratia_Plasmids_selfblastresults.txt | awk '$4>=99' | awk '{print $1"#"$2"#"$3"\t"$9-1"\t"$10"\t"$15"\t"$16}' | sort -k1,1 -k2,2n | mergeBed -c 4,5 -o distinct,distinct | awk '{print $0 "\t" ($3-$2)}' | groupBy -g 1 -c 4,5,6 -o distinct,distinct,sum | awk '{print $0 "\t" ($NF*100)/$2 "\t" ($NF*100)/$3}' | sort -nk5,5 -nk6,6 | awk '$NF>=99 && $(NF-1)>=99' | sed 's/#/\t/g' | awk '!($2==$3)' | cut -f2- >Serratia_Plasmids_selfBlastResults_QCov_Scov99_Id99.txt

# $ awk '{print $1 "\t" $2}' Serratia_Plasmids_selfBlastResults_QCov_Scov99_Id99.txt >Serratia_Plasmids_selfBlastResults_QCov_Scov99_Id99_pairs.txt

# All 9 plasmids are connected in Single component

# -----Cluster:1-------
# C1239_4_len_68647_circ_plasmid
# C392_4_len_68222_circ_plasmid
# C1292_6_len_67445_circ_plasmid
# C1379_4_len_67427_circ_plasmid
# C492_4_len_67445_circ_plasmid
# C1347_5_len_67445_circ_plasmid
# C1288_4_len_68706_circ_plasmid
# C1401_4_len_68706_circ_plasmid
# C1363_5_len_68645_circ_plasmid

# Since longest, let us take C1401_4_len_68706_circ_plasmid and check in CPE transmission dataset if it exists

# time blastn -db /storage/data/DATA4/analysis/25_5year_PlasmidEvolution/1071_Plasmids_BlastDB/Combined_1071_Circ_Plasmids.fasta.DB -query C1401_4_len_68706_circ_plasmid.fa -outfmt '6 qseqid sseqid pident nident length mismatch gapopen qstart qend sstart send evalue bitscore qlen slen' -out Serratia_Plasmids_C1401_4_len_68706_circ_vs_CPE_Tranmission_1071_Circ_Plasmids_blastresults.txt -num_threads 16 

# Filtering with 99% cov and 99% identity
# $ awk '{print FILENAME"\t"$0}' Serratia_Plasmids_selfblastresults.txt | awk '$4>=99' | awk '{print $1"#"$2"#"$3"\t"$9-1"\t"$10"\t"$15"\t"$16}' | sort -k1,1 -k2,2n | mergeBed -c 4,5 -o distinct,distinct | awk '{print $0 "\t" ($3-$2)}' | groupBy -g 1 -c 4,5,6 -o distinct,distinct,sum | awk '{print $0 "\t" ($NF*100)/$2 "\t" ($NF*100)/$3}' | sort -nk5,5 -nk6,6 | awk '$NF>=99 && $(NF-1)>=99' | sed 's/#/\t/g' | awk '!($2==$3)' | cut -f2- >Serratia_Plasmids_selfBlastResults_QCov_Scov99_Id99.txt

# But looks like it is better we use full deduped list of fasta

# $ time blastn -db /storage/data/DATA4/analysis/25_5year_PlasmidEvolution/1071_Plasmids_BlastDB/Combined_1071_Circ_Plasmids.fasta.DB -query Serratia_43_plasmids_dedup.fasta -outfmt '6 qseqid sseqid pident nident length mismatch gapopen qstart qend sstart send evalue bitscore qlen slen' -out Serratia_Plasmids_vs_CPE_Tranmission_1071_Circ_Plasmids_blastresults.txt -num_threads 64

# awk '{print FILENAME"\t"$0}' Serratia_Plasmids_vs_CPE_Tranmission_1071_Circ_Plasmids_blastresults.txt | awk '$4>=99' | awk '{print $1"#"$2"#"$3"\t"$9-1"\t"$10"\t"$15"\t"$16}' | sort -k1,1 -k2,2n | mergeBed -c 4,5 -o distinct,distinct | awk '{print $0 "\t" ($3-$2)}' | groupBy -g 1 -c 4,5,6 -o distinct,distinct,sum | awk '{print $0 "\t" ($NF*100)/$2 "\t" ($NF*100)/$3}' | sort -nk5,5 -nk6,6 | awk '$NF>=99 && $(NF-1)>=99' | sed 's/#/\t/g' | awk '!($2==$3)' | cut -f2-

# C1379_4_len_67427_circ_plasmid	batch1_08032019_ENT1213_Ctg3_len_67445_depth_4.02x_circ_blaOXA-48	67427	67445	67427	100	99.9733
# C1379_4_len_67427_circ_plasmid	batch1_08032019_ENT1214_Ctg21_len_67445_depth_2.25x_circ_blaOXA-48	67427	67445	67427	100	99.9733
# C1379_4_len_67427_circ_plasmid	batch20_11102019_ENT1240_Ctg4_len_67445_depth_1.14x_circ_blaOXA-48	67427	67445	67427	100	99.9733
# C1292_6_len_67445_circ_plasmid	batch1_08032019_ENT1213_Ctg3_len_67445_depth_4.02x_circ_blaOXA-48	67445	67445	67445	100	100
# C1292_6_len_67445_circ_plasmid	batch1_08032019_ENT1214_Ctg21_len_67445_depth_2.25x_circ_blaOXA-48	67445	67445	67445	100	100
# C1292_6_len_67445_circ_plasmid	batch20_11102019_ENT1240_Ctg4_len_67445_depth_1.14x_circ_blaOXA-48	67445	67445	67445	100	100
# C1347_5_len_67445_circ_plasmid	batch1_08032019_ENT1213_Ctg3_len_67445_depth_4.02x_circ_blaOXA-48	67445	67445	67445	100	100
# C1347_5_len_67445_circ_plasmid	batch1_08032019_ENT1214_Ctg21_len_67445_depth_2.25x_circ_blaOXA-48	67445	67445	67445	100	100
# C1347_5_len_67445_circ_plasmid	batch20_11102019_ENT1240_Ctg4_len_67445_depth_1.14x_circ_blaOXA-48	67445	67445	67445	100	100
# C492_4_len_67445_circ_plasmid	batch1_08032019_ENT1213_Ctg3_len_67445_depth_4.02x_circ_blaOXA-48	67445	67445	67445	100	100
# C492_4_len_67445_circ_plasmid	batch1_08032019_ENT1214_Ctg21_len_67445_depth_2.25x_circ_blaOXA-48	67445	67445	67445	100	100
# C492_4_len_67445_circ_plasmid	batch20_11102019_ENT1240_Ctg4_len_67445_depth_1.14x_circ_blaOXA-48	67445	67445	67445	100	100
# C392_4_len_68222_circ_plasmid	batch1_08032019_ENT1213_Ctg3_len_67445_depth_4.02x_circ_blaOXA-48	68222	67445	68222	100	101.152
# C392_4_len_68222_circ_plasmid	batch1_08032019_ENT1214_Ctg21_len_67445_depth_2.25x_circ_blaOXA-48	68222	67445	68222	100	101.152
# C392_4_len_68222_circ_plasmid	batch20_11102019_ENT1240_Ctg4_len_67445_depth_1.14x_circ_blaOXA-48	68222	67445	68222	100	101.152

# Yes! Serratia plasmids are in the CPE Transmission plasmid database


# Plasmid Phylogeny of Serratia and Klebsiella
# Took all the deduplicated plasmids fasta from both species and ran a parsnp. Not running gubbins as of now.



```

```{r OXA-48_plasmid_tree1, fig.width=12, fig.height=8, message=FALSE}

library(ape)
library(RCurl)

setwd("/data02/Analysis/Projects/8_Aqueos_samples/12_Plasmids_length_distr_by_species")

list.files()

# a tree was made using code in a previous blog post. 
# see here: http://rforbiochemists.blogspot.co.uk/2017/01/visualizing-kinome-in-r-simple-tree.html
# it's here on github
#link <- "https://raw.githubusercontent.com/brennanpincardiff/RforBiochemists/master/phyloTrees/kinaseTree_20161221"

# this will download it into R. 
# using read.tree() from ape package
PlotApePhyloTree <- function(newickFile, organism, width=30, height=30){
dev.new()
tree <- read.tree(file = newickFile)

# this tree looks quite nice in my opinion and is the starting point of this blog 
# plot(tree, "u", 
#      use.edge.length = FALSE,
#      show.tip.label = FALSE,
#      edge.color = "red")
# but lacking colours for the groups and any labels...





# it seems more traditional to draw trees from left to right. 
# plot(tree, 
#      use.edge.length = FALSE,
#      show.tip.label = FALSE,
#      edge.color = "red")


# with some names... this is slow to draw due to the names 
# plot(tree, 
#      use.edge.length = FALSE,
#      edge.color = "red",
#      cex = 0.25)



# and you can't read the names because there are >500




# to customise this tree in a way we want we need to understand a little more about trees
# we can find out more about an object by writing the name
tree
# "Phylogenetic tree with 516 tips and 514 internal nodes"

# by using the class() function
class(tree)
# "phylo"

# or by using the str() structure function
str(tree) 
# "List of 4"
# this list includes $edge, $Nnode, $ tip.label and $edge.length
# the tree$tip.label includes family designation
tree$tip.label  # 516 of these

# from the Science paper, we have seven kinase families:
# kinase categories... TK, TKL, STE, CK1, AGC, CAMK, CMGC
# with the following colours
# "red", "green", "paleblue", "orange", "yellow", "purple", "pink", "green" 


# by using the grep()function on the tree$tip.label part of the object
# we can find the tip labels that include "TK/" - i.e. tyrosine kinases
grep("_C1", tree$tip.label)  # gives a list of numbers with "TK/" in tip label
length(grep("_C1", tree$tip.label))
# thus there are 94 tip labels with that are designated TK (not TKL tyrosine kinase like)

# make a vector for each tip.label called tipcol with black on all of these...
tipcol <- rep('black', length(tree$tip.label))
tipcol
# make a vector with our list of kinase categories
kinaseCats <- c("Kl", "Se")

# make a vector of color we want:
colorsList <-c("red", "darkgreen", "blue", "orange", "navyblue", "purple", "maroon", "green")

# replace colours where grep gives "TK" as red, etc in a loop
for(i in 1:length(kinaseCats)){
  tipcol[grep(kinaseCats[i], tree$tip.label)] <- colorsList[i]
}

# plot with edge length false to see nodes better
#plot(tree)

# plot(tree, 
#      tip.color=tipcol, 
#      cex = 0.75)

#plot.phylo(tree,tip.color=tipcol,cex = 0.75,edge.width = 2)

#plot.phylo(tree,tip.color=tipcol,cex = 0.75, main = "Serratia marcescens Gubbins Tree", font=2)

plot.phylo(tree,
           tip.color=tipcol,
           cex = 0.75,
         #  use.edge.length = TRUE,
           main = paste0(organism," Gubbins Phylogenetic Tree"), 
           font=2)

library(extrafont)
#font_import() # import fonts - takes some time
loadfonts()
#png("phylo1.png",width = 800, height = 800)
pdf(paste0(organism,"_Gubbins_phylotree.pdf"), family="Times New Roman", width=width, height=height)

plot.phylo(tree,
           tip.color=tipcol,
           cex = 0.3,
           use.edge.length = TRUE,
           main = paste0(organism," ParSNP Phylogenetic Tree"), 
           font=2)
dev.off()

#graphics.off()

}

PlotApePhyloTree("parsnp.tree", "S.marcescens & K.pneuomoniae", width=5, height=5)



```

### Step 13: BEAST data analysis

We are going to provide the constant sites in BEAST xml file, so we will not append the sites to fasta sequences since this will take a long time. Rather, we provide SNP alignment (generated by Gubbins after removing recombination bases) to BEAST with constant sites in XML file.


```{r beast, engine = 'bash', eval = FALSE}
cd /storage/data/DATA4/analysis/27_Paeruginosa_183_samples_Shawn_with_16_Envsamples_31_Jeanette_40_jocelynsamples/9_SNP_Validation/PAE_266samples/Gubbins

#removed 5 samples without dates

C02099	  TTSH
C02123		TTSH
C02360		TTSH
C02433		TTSH
C02481		TTSH


$ grep '>' postGubbins.filtered_polymorphic_sites_copy.fasta | head
# >C1142
# >C1148
# >C1236
# >C1237
# >C1238
# >C1239
# >C1268
# >C1284
# >C1285
# >C1288

# Include decimal dates in a file called "IsolateDOC.txt" in the present directory. It should look like this:

# WGS_ID 	BEAST_header
# C17	C17|2019.4602739726
# C37	C37|2019.4602739726
# C94	C94|2019.47945205479
# C327	C327|2019.55616438356
# C392_Supp	C392_Supp|2019.55616438356
# C444_Supp	C444_Supp|2019.57534246575
# C468_Supp	C468_Supp|2019.57534246575
# C492_Supp	C492_Supp|2019.57534246575

# Now run perl script with postGubbins SNP file as input and provide prefix of the output file as below
perl /storage/apps/SNP_Validation_Scripts/tools/BEAST_DatesHeader.pl postGubbins.filtered_polymorphic_sites_copy.fasta postGubbins.filtered_polymorphic_sites

# Headers are now changed as required by BEAST
$ grep '>' postGubbins.filtered_polymorphic_sites_BEAST_withDates.fasta
# >C1142|2019.84383561644
# >C1148|2019.84383561644
# >C1236|2019.84383561644
# >C1237|2019.84383561644
# >C1238|2019.84383561644

grep -A1 '|' postGubbins.filtered_polymorphic_sites_BEAST_withDates.fasta | sed 's/--//g' | sed '/^$/d' >postGubbins.filtered_polymorphic_sites_BEAST_withDates_woSampleswithdates.fasta

# First, we need to remove the outlier samples using tempest by plotting root-to-tip distance vs time correlation. For this we need to run 
# iqtree first which generates a tree

time iqtree2 -s postGubbins.filtered_polymorphic_sites_BEAST_withDates.fasta -T AUTO

load .treefile into tempest # Root-to-tip is positively correlated

# Now load this in BEAUTi to generate input XML for BEAST

Partitions Tab: Open beauti > File Import Alignment (No change in parititons) 
Tip Dates Tab: Click use tip dates > Auto-configure > use_everything after "|"> Ok
Site Model Tab: BEAST Model Test
Clock Model Tab: Strict Model 
Priors Tab: Tree > Coalescent Constant Population (Rest of the parameters: default)
MCMC Tab: 10000000
File > save XML file

# Now to append constant sites, we need to calculate the number of constant sites that we want to append

# Sometime, faidx does not work, if there are any other characters like dash ("-"). SO remove them
$ sed -i 's/-/N/g' postGubbins.filtered_polymorphic_sites_BEAST_withDates.fasta 

$ faidx -i nucleotide NC_002516.fasta -o NC_002516.fasta_faidx

$ faidx -i nucleotide postGubbins.filtered_polymorphic_sites_BEAST_withDates.fasta -o postGubbins.filtered_polymorphic_sites_BEAST_withDates_faidx

# These two files are passed to a perl script which will calculate the invariant sites to put in

perl /storage/apps/SNP_Validation_Scripts/tools/calculate_InvariantSites.pl NC_002516.fasta_faidx postGubbins.filtered_polymorphic_sites_BEAST_withDates_faidx 

# check this folder for the correct XML used: /data02/Analysis/Projects/6_Paeruginosa_199_TTSH_31_NUH_40_SGH/SNIPPY_withoutSNPvalidation
# The above script generates this output

A:1056134,T:1038950,C:2102687,G:2066633,N:0 # Ref genome nucl stats
A:0.168592894072604,T:0.165849775972303,C:0.335656352942754,G:0.329900977012338 # Ref genome nucl proportions
---->Reference Genome nucleotide proportion A:0,T:0,C:0,G:0
core genome end: 544 # SNP alignment total bases
Reference genome size (end) is:6264404, core genome size (end): 544, Difference: 6263860 # Difference is total invariant sites


A:1056042.28546562,T:1038859.77772187,C:2102504.402944,G:2066453.53386851 # According to above proportion, the difference calculated above is distributed
 # Decimals to integer
6263860==6263860 ----> rounding is perfect! # Check if the integer generated above add up to "Difference" we calculated above


A:1056042,T:1038860,C:2102504,G:2066454 # Add this sites to the BEAST XML generated by Beauti

# Now open XML file in text editor
Step 1: Change `<data id="postGubbins" name="alignment">` to `<data id="coreOriginal" name="alignment">`
Step 2: After </data> tag, we need to add another tag
Step 3: The invariant sites we calculated above should be pasted in the constantSiteWeights parameter with order of A-C-G-T.
So the final line should be:
<data id="postGubbins" spec="FilteredAlignment" filter="-" data="@coreOriginal" constantSiteWeights="1040770 1522977 1520300 1029706"/> 

# or simply run the script by providing xml file and A-C-G-T nulceotide we got from above calculate_InvariantSites.pl script
perl addConstSites2_BeautiXML.pl postGubbins.filtered_polymorphic_sites_BEAST_withDates_10M.xml 1056042 2102504 2066454 1038860

# Thats it! Lets Load in the BEAST. 

time beast -beagle -beagle_GPU postGubbins.filtered_polymorphic_sites_BEAST_withDates_10M.xml

# To run multiple runs, I wrote a short script to run multiple BEAST runs at one go.

time bash ./multiple_Run_BEAST.sh &

```

### Step 14a: Analysing BEAST results

- Once BEAST run has finished, then load the .log files into tracer.
- If the multiple runs independently has ESS> 200 and Tracer looks okay, then we can select the clockRate and copy the 95% HPD interval.
- In the case of P.aeruginosa 95% HPD interval	is 1.9651E-3 - 2.8706E-3 - This is in the units SNPs/Site/Year
- Now, we need to calculate the SNPs/genome/year - for this we multiply the rate with it's genome size 
- This translates to 12310.1803004 - 17982.5981224 SNPs/genome/year for P.aeruginosa species
- Now, let us combine the logs using logcombiner  - took less than a minute to run this - 
-  Now, let us combine the trees using logcombiner - took less than a minute to run this - Rerun this step with resample states at lower frequency if the treeannotater is crashing eg: 10000 - (In my case, since I have a bigger dataset, I have set resample states at lower frequency: 10000 in the logcombiner when combining trees)
- use Treeannotater to annotate tree to less than 10 minutes with 10000 sample frequency
  - Burnin percentage: 10
  - Target tree type: Maximum credibility tree
  - Node heights: Common ancestor heights
  - choose low memory
  
# This will help find the values 2020.8524590163911.873=2008.9, 2020.8524590163919.252= 2001.6



### Step 14b: Plotting substitution rates for each species using R

```{r substitutionrates, fig.width=16, fig.height=18, message=FALSE}





```


### Step 15: Running Poisson Script to find the Transmission Plausible Isolate Pairs

We use `dnaDist_Poisson_withConnectedComp.R` script. Modify script wherever you see "###Change".
Once you Run the R Script, you will generate "Samples_ClusterNumber.txt".

### Additional analysis - Checking presence of specific genomic region (ICEtn in this case)

Rename the files to SRST2 compatible file names

```{r srst2, engine = 'bash', eval = FALSE}
for d in $(ls -l ../2_CleanData/ | grep -v '>' | grep 'C' | awk '{print $NF}'); do ln -s /storage/data/DATA4/analysis/27_Paeruginosa_183_samples_Shawn_with_16_Envsamples/2_CleanData/$d/*_1_bbmap_adaptertrimmed.fq "$d"_S1_L001_R1_001.fastq; done
for d in $(ls -l ../2_CleanData/ | grep -v '>' | grep 'C' | awk '{print $NF}'); do ln -s /storage/data/DATA4/analysis/27_Paeruginosa_183_samples_Shawn_with_16_Envsamples/2_CleanData/$d/*_2_bbmap_adaptertrimmed.fq "$d"_S1_L001_R2_001.fastq; done


for d in $(ls *.fastq | cut -f1 -d "_"  | sort -u); do mkdir $d; mv "$d"_*.fastq $d; done

time for d in $(ls -d */ | tr -d "/"); do echo $d; cd $d;  R1=`ls *_S1_L001_R1_001.fastq`; R2=`ls *_S1_L001_R2_001.fastq`; echo "R1: $R1----R2: $R2"; time srst2 --input_pe $R1 $R2 --output test_srst2 --log --gene_db /storage/data/DATA4/analysis/27_Paeruginosa_183_samples_Shawn/16_Check_presence_of_ICETn43716385/ICETn43716385.fasta --threads 48; cd ..; done

$ cat */test_srst2__fullgenes__ICETn43716385__results.txt | awk '{print $1 "\t" $2 "\t" $4 "\t" $5 "\t" $7}' | grep -v 'coverage' | column -t
```

### Optional - Testing

```{r testing_only_snippy_results, engine = 'bash', eval = FALSE}
cd /storage/data/DATA4/analysis/27_Paeruginosa_183_samples_Shawn_with_16_Envsamples_31_Jeanette_40_jocelynsamples/9_SNP_Validation/PAE_266samples/snippy

$ snippy-core --noref 345 544 569 571 586 587 590 592 596 597 599 624 629 642 653 680 681 683 688 693 694 699 700 701 710 723 725 728 730 733 734 C00024 C00027 C00042 C00056 C00068 C00073 C00118 C00130 C00131 C00132 C00136 C00177 C00183 C00186 C00194 C00219 C00220 C00270 C00280 C00281 C00309 C00318 C00336 C00337 C00369 C00386 C00387 C00389 C00391 C00400 C00401 C00432 C00461 C00472 C00476 C00487 C00488 C00496 C00499 C00507 C00569 C00584 C00625 C00651 C00669 C00696 C00703 C00707 C00716 C00718 C00746 C00765 C00787 C00811 C00814 C00818 C00819 C00851 C00881 C00916 C00950 C00954 C01074 C01076 C01094 C01107 C01121 C01148 C01157 C01187 C01196 C01201 C01206 C01237 C01238 C01243 C01244 C01245 C01248 C01267 C01295 C01305 C01336 C01337 C01338 C01339 C01340 C01390 C01391 C01395 C01399 C01412 C01429 C01446 C01447 C01448 C01494 C01495 C01505 C01528 C01571 C01572 C01578 C01591 C01603 C01623 C01657 C01681 C01683 C01695 C01716 C01727 C01747 C01756 C01758 C01776 C01809 C01812 C01813 C01835 C01843 C01844 C01845 C01863 C01871 C01872 C01873 C01877 C01883 C01888 C01889 C01890 C01907 C01914 C01921 C01922 C01958 C01965 C01978 C01992 C01996 C02021 C02071 C02091 C02092 C02099 C02120 C02123 C02164 C02170 C02191 C02223 C02224 C02247 C02252 C02263 C02288 C02318 C02322 C02329 C02340 C02345 C02349 C02357 C02358 C02360 C02361 C02364 C02374 C02382 C02385 C02397 C02429 C02430 C02431 C02432 C02433 C02434 C02481 C1000 C1042 C1044 C1094 C1199 C498 C503 C509 C529 C545 C581 C592 C616 C636 C717 C779 PA0197 PA0260 PA0312 PA0315 PA0317 PA0349 PA0360 PA0393 PA0395 PA0408 PA0424 PA0436 PA0453 PA0461 PA0465 PA0472 PA0473 PA0480 PA0486 PA0556 PA0823 PA0834 PA0874 PA0887 PA0895 PA0900 PA0939 PA0972 PA1030 PA1111 PA1250 PA1284 PA1293 PA1301 PA1318 PA1350 PA1352 PA1353 PA1379 PA1388

/storage/apps/snippy/bin/snippy-clean_full_aln core.full.aln > clean.core.full.aln

 time run_gubbins.py -p gubbins clean.core.full.aln --threads 48 --verbose



```
