## Molecular epidemiology and genomic analysis of XX isolates and XX reveals transmission of bacterial plasmids harboring carbapenem resistant genes

[RefPaper1](https://sci-hub.se/https://doi.org/10.1093/jac/dkaa398) , [RefPaper2](https://sci-hub.se/https://doi.org/10.1093/jac/dkz146)

### Introduction:
* Most important problem in healthcare setting
* Which order of bacteria is causing it? What makes it more deadly and powerful to attack the humans?
* CPE - mostly which organisms cause the frequent attacks?
* What causes the Carbapenemase to move? Transmissible mobile genetic elements
* Which sepcies in this study we are concentrating about?
* Are there any prevalent ST clone in our dataset?
* Treatment of antibiotics
* Acquisition of resistance genes through mobile genetic elements
    - specific genomic islands 
    - transposons
    - plasmids
    - integrative and conjugative elements
    - integrons
* Severe infection due to virulence factors
* Toxins of the type III secretion system to circumvent the host immune system and establish infection
* Of the four exotoxins (ExoS, ExoT, ExoU, ExoY), ExoU, a potent phospholipase that disrupts the plasma membrane and leads to rapid cell death, is the most virulent 

### Methods

### Results

* **Bacterial isolates**
  - From what time to what time these isolates have been collected?
  - Out of the collected, how many have carbapenemases?
  - Where are isolates collected from patients? Which body sites? If environmental, which surrounding sites? Give the count of each site.
  - Out of this, what bacterial species was found and the count of each species.
  - Is there any trend of increasing in the number of isolates by species by year?

  
* **Location of CP genes**
  - Is it chromsomosomal or plasmids?
  - If plasmids what is the size of the plasmids

* **WGS analysis**
   - Genome Characteristics: The average genome size of the de novo assembled genomes for each species - Information about all genome characteristics is summarized in Table. 
   - Resistome analysis - what genes are conferring resistance to aminoglycosides (aacA4 and aadA1), macrolides (mphE and msrE), chloramphenicol (catA1 and catB2), sulphonamides (sul1) and trimethoprim/sulfamethoxazole (dfrB1) in each of the species isolates. In OXA-48-CP isolates, only one gene involved in aminoglycoside antimicrobial resistance was found (aacA4) (Figure S3). Show the facet grid or PCA plot for the resistome
   - Type of plasmid reconstructions showed in each of the 4 species - what is the size of the plasmid with CP gene ? Is it available in NCBI? If not, what is the coverage and identity? Is this the same plasmid in the other species as well carrying the CP gene? What are other plasmids in other species? and their details?
   - Phylogenetic analysis of all the genomes for each species 
   - Typing, plasmid, and genomic transmission analysis - What does MLST analysis reveal? How many groups of STs identified for each species? 
   - Plasmid replicon types were detected, and associations with carbapenemase genes were assessed - Among OXA-48-producing isolates, Inc-- replicon plasmids were common in Serratia marcescens (85.7%) and K. pneumoniae (61.1%), IncL/M plasmids were more common among E. coli and Enterobacter spp. (66.7% and 50.0%, respectively), and IncHI2 plasmids were most common among E. cloacae complex (61.5%). All three types have previously been described to carry the IMP-4 gene in Australia (9, 34). KPC-2-producing K. pneumoniae commonly carried IncFIB (94.6%), ColRNAI (90.2%), and IncX3 (85.7%). In contrast, NDM and OXA isolates carried a wide range of plasmid types. Plasmid replicon sequences were only rarely detected on the same contig as a carbapenemase gene. Fifteen isolates had no plasmid replicon sequences detected at these identity/coverage cutoffs, including all IMI- and SME-producing isolates (chromosomal carbapenemases).
   - Genomic transmission analyses were performed on 33 subgroups of the same species, carbapenemase gene, and ST (where available), including 131 isolates from 119 patients; subgroups included 2 to 14 isolates (median, 3 isolates) (Table S6). Pairwise SNP distances were plotted for each species
   - Plasmid replicon typing results were consistent with transmission analysis results in epidemiologically confirmed clusters, in that all isolates within a cluster carried the same plasmid replicon types, with the exception of the mixed NDM-5/NDM-5 plus OXA-232 cluster (K. pneumoniae ST16), where only the isolates with OXA-232 had the ColKP3-type plasmid detected, suggesting carriage of the OXA-232 gene on this plasmid.
   -  Not much diversity of resistance genes in ST308 isolates - Spread of both of these lineages are more or less heavily dependent on these resistant genes

### Discussion
   - OXA-48-producing Enterobacterales isolates have been reported from hospital and extra-hospital reservoirs. Most of the studies on
this subject have been conducted in developed countries, but the major epicentres of OXA-48-producing Enterobacterales are located in Mediterranean countries such as Algeria, where this resistance mechanism is endemic.
   - Therefore our study encompasses OXA-48 isolates - Plasmid mediated transmission - usual species but serratia - associated with specific plasmids and replicons
   - Along these lines, the presence of additional antimicrobial co-resistance genes could have played a key role in the selection and persistence of this IncL-pVIM-1 variant over the conserved IncL-pOXA-48.

### Conclusions
 
 - In conclusion, how many lineages combined with versions of the same plasmid carrying XX cp genes or OXA-48 have contributed to the dispersion of CPSm in our institution.
 - Moreover, our results also indicate that the IncL-pOXA-48 plasmid has the ability not only to transfer among different Enterobacterales species, including
S. marcescens, but also to exchange genetic material. This fact facilitates the acquisition of other genetic structures carrying multiple resistance genes, including other carbapenemases such as blaVIM-1, thus generating nosocomial pathogens with a higher MDR profile. The transmission of these plasmids among different clinical pathogens including S. marcescens clones with the potential capacity to spread in hospital units with a high risk of infection,
such as ICUs, represents a new challenge and should be considered a public health priority. 
This study reinforces the relevance of species different from K. pneumoniae or E. coli in the CPE landscape and circulating lineages and plasmids in the local
CPE epidemiology.
